{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TODO (project-level): \n",
    "\n",
    "- Generation of Cell names\n",
    "- names for higher subclasses\n",
    "- penotypic & Genotypic -- Check threshold\n",
    "  \n",
    "----------------------------------------------------\n",
    "\n",
    "- Start building from individual neurons\n",
    "- Naming scheme: Tasic 15, cluster [cluster_number]\n",
    "- Function that changes names flexibly\n",
    "- Templates & References\n",
    "    - Bolser Lewis \n",
    "    - OntTerm triplesimple\n",
    "    - nistd/core.py\n",
    "    - curation.py = triplesExport\n",
    "    \n",
    "#### Code-level:\n",
    "- Start using templates from NeuronLangExample\n",
    "- Recapitulating the matrix with definitive phenotypes\n",
    "- populate namespace with definitive phenotypesi\n",
    "\n",
    "Definitive phenotype:\n",
    "    - Cre line [NCBI Taxon] [Allen Brain Institute]\n",
    "    - V1 \n",
    "    - Layer of dissection [UBERON]\n",
    "\n",
    "Inferred gene markers phenotype\n",
    "    - Cluster ID [ILX]\n",
    "    - Genes present [NCBIGene]\n",
    "    - Genes Absent [NCBIGene]\n",
    "\n",
    "#Tasic hiearchy\n",
    "### For the Future:\n",
    "- Continuous Phenotypes (gene counts, etc)\n",
    "- Phenotype Inheritance in subclasses?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "http://casestudies.brain-map.org/celltax\n",
    "\n",
    "https://www.nature.com/articles/nn.4216"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "#standard libraries\n",
    "from pathlib import Path\n",
    "from importlib import reload\n",
    "\n",
    "#extended libraries\n",
    "import requests\n",
    "import rdflib\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy as sp\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "#tree stuff\n",
    "import anytree\n",
    "import anytree.util\n",
    "from anytree.importer import DictImporter\n",
    "from anytree.exporter import DotExporter\n",
    "\n",
    "#custom imports\n",
    "from neurondm import *\n",
    "from neurondm.models.huang2017 import Genes\n",
    "from pyontutils.namespaces import ilxtr as pred\n",
    "from neurondm import phenotype_namespaces as phns\n",
    "from pyontutils.closed_namespaces import rdf, rdfs, owl\n",
    "#from nifstd.nifstd_tools.utils import ncbigenemapping\n",
    "from utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cellguard(addns=False):\n",
    "    # unfortunately ipy.hooks['pre_run_code_hook'].add(__cellguard)\n",
    "    # causes this to be called too frequently :/\n",
    "    setLocalNames()\n",
    "    setLocalContext()\n",
    "    if addns:\n",
    "        setLocalNames(phns.BBP)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_clf = pd.read_csv('Data/cell_classification.csv')\n",
    "df_cls_mtd = pd.read_csv('Data/cluster_metadata.csv')\n",
    "df_cell_mtd = pd.read_csv('Data/cell_metadata.csv')\n",
    "\n",
    "df_cre_mtd  = pd.read_excel('Data/tasic_crelines.xlsx')\n",
    "\n",
    "dendro = pd.read_csv(\"Data/web/big_tree_final.csv\", dtype = {'position':str})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Information about the cluster membership of each cell,\n",
    "including whether the cell is a \"core\" (unambiguously assigned to a single cluster) \n",
    "or \"transition\" (shares membership between two clusters) cell, \n",
    "as well as its membership score (from 0-10) for each cluster (labeled f01 to f49). \n",
    "\"\"\"\n",
    "df_clf.head()\n",
    "#TODO: Only including unambiguous cells with f-values = 10 for one cluster only."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    " Information about each data-driven cluster, including its label, \n",
    " the corresponding label in Tasic et al. (Nat. Neuro, 2106), the primary cell class \n",
    " membership, and marker genes (including genes with widespread expression in the cluster, \n",
    " sparse expression in the cluster, and no expression in the cluster). \n",
    "\"\"\"\n",
    "df_cls_mtd.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Information about each cell profiled, including its nomenclature, \n",
    "Cre line of origin, dissection, date of collection and sequencing, \n",
    "and read mapping statistics\n",
    "\"\"\"\n",
    "df_cell_mtd.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_clf['coretype'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cell_mtd['layer_dissectoin'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cell_mtd['cre'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_clf.rename(index=str, columns={\"Unnamed: 0\":\"cell_index\"}, inplace = True)\n",
    "df_clf = df_clf[['cell_index','coretype', 'primary', 'secondary']]\n",
    "\n",
    "#Reorder and drop columns that are not interesting\n",
    "df_cell_mtd = df_cell_mtd[['short_name','cre','major_class','sub_class',\n",
    "                           'major_dissection', 'layer_dissectoin']]\n",
    "df_cls_mtd = df_cls_mtd[['cluster_id','cluster_order','vignette_label',\n",
    "                         'group','markers_present','markers_sparse',\n",
    "                         'genes_absent','Tasic_et_al_2016_label']]\n",
    "\n",
    "df_cls_mtd['cluster_id'] = df_cls_mtd['cluster_id'].astype(str)\n",
    "\n",
    "#merge\n",
    "df_types = df_cell_mtd.merge(df_clf, left_on='short_name', right_on='cell_index')\n",
    "df_types = df_types[['short_name', 'coretype', 'primary', 'secondary',\n",
    "                     'cre','major_dissection', 'layer_dissectoin']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cre line conversion\n",
    "df_cre_mtd.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "df_types.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# metada query test\n",
    "df_cre_mtd[df_cre_mtd['Abbreviation'] == df_types['cre'][1000]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "main dataframe: df_types\n",
    "\n",
    "metadata: cre_mtd & cls_mtd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dendrogram Parsing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cd34</td>\n",
       "      <td>000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ndnf Car4</td>\n",
       "      <td>0000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ndnf Cxcl14</td>\n",
       "      <td>0000011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Vip Gpc3</td>\n",
       "      <td>0000100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Vip Chat</td>\n",
       "      <td>0000101</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       cluster position\n",
       "0         Cd34   000000\n",
       "1    Ndnf Car4  0000010\n",
       "2  Ndnf Cxcl14  0000011\n",
       "3     Vip Gpc3  0000100\n",
       "4     Vip Chat  0000101"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dendro.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "tree_dict = {'label':'root'} #base dictionary\n",
    "\n",
    "#parses binary node positions into a dictionary with tree structure\n",
    "for label, bin_str in zip(dendro['cluster'], dendro['position']):\n",
    "    parse_binary(tree_dict, bin_str, label)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnyNode(label='root', name='1', pos=1)\n",
      "├── AnyNode(name='2', pos=2)\n",
      "│   ├── AnyNode(name='4', pos=4)\n",
      "│   │   ├── AnyNode(name='8', pos=8)\n",
      "│   │   │   ├── AnyNode(name='16', pos=16)\n",
      "│   │   │   │   ├── AnyNode(name='28', pos=28)\n",
      "│   │   │   │   │   ├── AnyNode(label='Cd34', name='42 Cd34', pos=42)\n",
      "│   │   │   │   │   └── AnyNode(name='43', pos=43)\n",
      "│   │   │   │   │       ├── AnyNode(label='Ndnf Car4', name='60 Ndnf Car4', pos=60)\n",
      "│   │   │   │   │       └── AnyNode(label='Ndnf Cxcl14', name='61 Ndnf Cxcl14', pos=61)\n",
      "│   │   │   │   └── AnyNode(name='29', pos=29)\n",
      "│   │   │   │       ├── AnyNode(name='44', pos=44)\n",
      "│   │   │   │       │   ├── AnyNode(label='Vip Gpc3', name='62 Vip Gpc3', pos=62)\n",
      "│   │   │   │       │   └── AnyNode(label='Vip Chat', name='63 Vip Chat', pos=63)\n",
      "│   │   │   │       └── AnyNode(name='45', pos=45)\n",
      "│   │   │   │           ├── AnyNode(name='64', pos=64)\n",
      "│   │   │   │           │   ├── AnyNode(label='Vip Cxcl14_Car4', name='80 Vip Cxcl14_Car4', pos=80)\n",
      "│   │   │   │           │   └── AnyNode(label='Vip Mybpc1', name='81 Vip Mybpc1', pos=81)\n",
      "│   │   │   │           └── AnyNode(label='Vip Sncg', name='65 Vip Sncg', pos=65)\n",
      "│   │   │   └── AnyNode(label='Sncg', name='17 Sncg', pos=17)\n",
      "│   │   └── AnyNode(name='9', pos=9)\n",
      "│   │       ├── AnyNode(name='18', pos=18)\n",
      "│   │       │   ├── AnyNode(name='30', pos=30)\n",
      "│   │       │   │   ├── AnyNode(label='Sst Chodl', name='46 Sst Chodl', pos=46)\n",
      "│   │       │   │   └── AnyNode(name='47', pos=47)\n",
      "│   │       │   │       ├── AnyNode(name='66', pos=66)\n",
      "│   │       │   │       │   ├── AnyNode(name='82', pos=82)\n",
      "│   │       │   │       │   │   ├── AnyNode(label='Sst Cdk6', name='88 Sst Cdk6', pos=88)\n",
      "│   │       │   │       │   │   └── AnyNode(name='89', pos=89)\n",
      "│   │       │   │       │   │       ├── AnyNode(label='Sst Cbln4', name='94 Sst Cbln4', pos=94)\n",
      "│   │       │   │       │   │       └── AnyNode(label='Sst Myh8', name='95 Sst Myh8', pos=95)\n",
      "│   │       │   │       │   └── AnyNode(label='Sst Tacstd2', name='83 Sst Tacstd2', pos=83)\n",
      "│   │       │   │       └── AnyNode(label='Sst Th', name='67 Sst Th', pos=67)\n",
      "│   │       │   └── AnyNode(name='31', pos=31)\n",
      "│   │       │       ├── AnyNode(label='Pvalb Cpne5', name='48 Pvalb Cpne5', pos=48)\n",
      "│   │       │       └── AnyNode(name='49', pos=49)\n",
      "│   │       │           ├── AnyNode(label='Pvalb Tpbg', name='68 Pvalb Tpbg', pos=68)\n",
      "│   │       │           └── AnyNode(name='69', pos=69)\n",
      "│   │       │               ├── AnyNode(name='84', pos=84)\n",
      "│   │       │               │   ├── AnyNode(label='Pvalb Gpx3', name='90 Pvalb Gpx3', pos=90)\n",
      "│   │       │               │   └── AnyNode(name='91', pos=91)\n",
      "│   │       │               │       ├── AnyNode(label='Pvalb Rspo2', name='96 Pvalb Rspo2', pos=96)\n",
      "│   │       │               │       └── AnyNode(label='Pvalb Wt1', name='97 Pvalb Wt1', pos=97)\n",
      "│   │       │               └── AnyNode(name='85', pos=85)\n",
      "│   │       │                   ├── AnyNode(label='Pvalb Obox3', name='92 Pvalb Obox3', pos=92)\n",
      "│   │       │                   └── AnyNode(label='Pvalb Tacr3', name='93 Pvalb Tacr3', pos=93)\n",
      "│   │       └── AnyNode(label='Igtp', name='19 Igtp', pos=19)\n",
      "│   └── AnyNode(name='5', pos=5)\n",
      "│       ├── AnyNode(name='10', pos=10)\n",
      "│       │   ├── AnyNode(name='20', pos=20)\n",
      "│       │   │   ├── AnyNode(label='L2 Ngb', name='32 L2 Ngb', pos=32)\n",
      "│       │   │   └── AnyNode(label='L2/3 Otof', name='33 L2/3 Otof', pos=33)\n",
      "│       │   └── AnyNode(name='21', pos=21)\n",
      "│       │       ├── AnyNode(name='34', pos=34)\n",
      "│       │       │   ├── AnyNode(name='50', pos=50)\n",
      "│       │       │   │   ├── AnyNode(label='L4 Sparcl1', name='70 L4 Sparcl1', pos=70)\n",
      "│       │       │   │   └── AnyNode(label='L4 Scnn1a', name='71 L4 Scnn1a', pos=71)\n",
      "│       │       │   └── AnyNode(label='L4 Arf5', name='51 L4 Arf5', pos=51)\n",
      "│       │       └── AnyNode(label='L4 Hsd11b1', name='35 L4 Hsd11b1', pos=35)\n",
      "│       └── AnyNode(name='11', pos=11)\n",
      "│           ├── AnyNode(name='22', pos=22)\n",
      "│           │   ├── AnyNode(name='36', pos=36)\n",
      "│           │   │   ├── AnyNode(name='52', pos=52)\n",
      "│           │   │   │   ├── AnyNode(label='L5a Fam5c', name='72 L5a Fam5c', pos=72)\n",
      "│           │   │   │   └── AnyNode(name='73', pos=73)\n",
      "│           │   │   │       ├── AnyNode(label='L5a Syt17', name='86 L5a Syt17', pos=86)\n",
      "│           │   │   │       └── AnyNode(label='L5a Batf3', name='87 L5a Batf3', pos=87)\n",
      "│           │   │   └── AnyNode(label='L5 Ucma', name='53 L5 Ucma', pos=53)\n",
      "│           │   └── AnyNode(name='37', pos=37)\n",
      "│           │       ├── AnyNode(name='54', pos=54)\n",
      "│           │       │   ├── AnyNode(label='L5b Samd3', name='74 L5b Samd3', pos=74)\n",
      "│           │       │   └── AnyNode(label='L5b Cdh13', name='75 L5b Cdh13', pos=75)\n",
      "│           │       └── AnyNode(label='L5 Chrna6', name='55 L5 Chrna6', pos=55)\n",
      "│           └── AnyNode(name='23', pos=23)\n",
      "│               ├── AnyNode(name='38', pos=38)\n",
      "│               │   ├── AnyNode(name='56', pos=56)\n",
      "│               │   │   ├── AnyNode(label='L6a Mgp', name='76 L6a Mgp', pos=76)\n",
      "│               │   │   └── AnyNode(label='L6a Plcxd3', name='77 L6a Plcxd3', pos=77)\n",
      "│               │   └── AnyNode(name='57', pos=57)\n",
      "│               │       ├── AnyNode(label='L6a Car12', name='78 L6a Car12', pos=78)\n",
      "│               │       └── AnyNode(label='L6a Syt17', name='79 L6a Syt17', pos=79)\n",
      "│               └── AnyNode(name='39', pos=39)\n",
      "│                   ├── AnyNode(label='L6b Trh', name='58 L6b Trh', pos=58)\n",
      "│                   └── AnyNode(label='L6b Mup5', name='59 L6b Mup5', pos=59)\n",
      "└── AnyNode(name='3', pos=3)\n",
      "    ├── AnyNode(name='6', pos=6)\n",
      "    │   ├── AnyNode(label='Astro Gja1', name='12 Astro Gja1', pos=12)\n",
      "    │   └── AnyNode(name='13', pos=13)\n",
      "    │       ├── AnyNode(label='OPC Pdgfra', name='24 OPC Pdgfra', pos=24)\n",
      "    │       └── AnyNode(name='25', pos=25)\n",
      "    │           ├── AnyNode(label='OPC/Oligo', name='40 OPC/Oligo', pos=40)\n",
      "    │           └── AnyNode(label='Oligo Opalin', name='41 Oligo Opalin', pos=41)\n",
      "    └── AnyNode(name='7', pos=7)\n",
      "        ├── AnyNode(label='Micro Ctss', name='14 Micro Ctss', pos=14)\n",
      "        └── AnyNode(name='15', pos=15)\n",
      "            ├── AnyNode(label='Endo Tbc1d4', name='26 Endo Tbc1d4', pos=26)\n",
      "            └── AnyNode(label='Endo Myl9', name='27 Endo Myl9', pos=27)\n"
     ]
    }
   ],
   "source": [
    "importer = DictImporter()\n",
    "tree = importer.import_(tree_dict)\n",
    "\n",
    "for ind, node in enumerate(anytree.LevelOrderIter(tree)):\n",
    "    node.pos = ind + 1 #node index starting at one\n",
    "    if node.is_leaf:\n",
    "        node.name = str(ind + 1) + \" \" + node.label\n",
    "    else:\n",
    "        node.name = str(ind + 1)\n",
    "        \n",
    "print(anytree.RenderTree(tree))\n",
    "DotExporter(tree).to_picture(\"dendro.png\")\n",
    "\n",
    "#node.name for rendering\n",
    "#node.pos & node.label used for parsing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "for leaf in tree.leaves: #TODO: phenotypes & deal with mismatching\n",
    "    #edge cases for the last two endothelial cells\n",
    "    if leaf.label  == 'Endo Tbc1d4':\n",
    "        leaf.cluster_id = 'f48'\n",
    "    elif leaf.label == 'Endo Myl9':\n",
    "        leaf.cluster_id = 'f49'\n",
    "    else:\n",
    "        leaf.cluster_id = list(df_cls_mtd[df_cls_mtd['vignette_label'] == leaf.label]['cluster_id'])[0]\n",
    "    \n",
    "    #FIXME: label name matching somehow not working\n",
    "    \"\"\"\n",
    "    genes_present = list(df_cls_mtd[df_cls_mtd['cluster_id'] == leaf.cluster_id]['markers_present'].str.split(\",\"))\n",
    "    genes_absent = list(df_cls_mtd[df_cls_mtd['cluster_id'] == leaf.cluster_id]['genes_absent'].str.split(\",\"))\n",
    "    if genes_present == np.nan:\n",
    "        leaf.genes_present = set()\n",
    "    if genes_absent == np.nan:\n",
    "        leaf.genes_absent = set()\n",
    "    else:\n",
    "        leaf.genes_present = set(genes_present[0])\n",
    "        #leaf.genes_absent = set(genes_absent)\n",
    "    \"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gene and cluster columns converted from dendrogram structure\n",
    "df_types['cluster'] = df_types.apply(cluster_converter, tree = tree, axis = 1)\n",
    "df_types['markers_present'] = df_types.apply(gene_merge, df = df_cls_mtd, index = 'markers_present', axis = 1)\n",
    "df_types['markers_absent'] = df_types.apply(gene_merge, df = df_cls_mtd, index = 'genes_absent', axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_types[df_types['coretype'] == 'Transition']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.value_counts(df_types[df_types['coretype'] == 'Transition']['cluster']).plot.bar()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phenotype Bagging"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Tables should now be in final version\n",
    "df_types.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cre_df3 = pd.merge(df_cre_mtd, cre_ref,  how='inner', \n",
    "                  left_on='Public Repository Stock #',right_on = 'stock_number')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cre_df3.drop_duplicates(inplace=True)\n",
    "cre_df3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cre_df2 = pd.merge(df_cre_mtd, cre_ref,  how='inner', \n",
    "                  left_on='Driver Line',right_on = 'name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "cre_df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cre_df = pd.merge(df_cre_mtd, cre_ref,  how='left', \n",
    "                  left_on=['Driver Line','Public Repository Stock #'], \n",
    "                  right_on = ['name','stock_number'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cre_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "response = requests.get('http://api.brain-map.org/api/v2/data/query.json?criteria='\n",
    "                        'model::TransgenicLine,rma::options[num_rows$eqall]')\n",
    "cre_ref = pd.DataFrame(response.json()['msg'])\n",
    "cre_ref['stock_number'] = pd.to_numeric(cre_ref['stock_number'])\n",
    "\n",
    "#merge both cre dataframes\n",
    "cre_df1 = pd.merge(df_cre_mtd, cre_ref,  how='inner', \n",
    "                   left_on='Driver Line',right_on = 'name')\n",
    "\n",
    "cre_df2 = pd.merge(df_cre_mtd.dropna(subset = ['Public Repository Stock #']), \n",
    "                   cre_ref.dropna(subset = ['stock_number']),  how='inner', \n",
    "                   left_on='Public Repository Stock #',\n",
    "                   right_on = 'stock_number')\n",
    "\n",
    "cre_df1.drop_duplicates(inplace = True)\n",
    "cre_df2.drop_duplicates(inplace = True)\n",
    "cre_df = pd.concat([cre_df1, cre_df2], axis = 0)\n",
    "cre_df.drop_duplicates(inplace = True)\n",
    "#get rid of redunant/useless columns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Rbp4              173\n",
       "Htr3a             123\n",
       "Cux2              122\n",
       "Sst               107\n",
       "Scnn1a-Tg3         99\n",
       "Pvalb              89\n",
       "PvalbF-Gad2        87\n",
       "Chrna2             84\n",
       "Gad2               77\n",
       "Ntsr1              77\n",
       "Nos1               72\n",
       "Vip                68\n",
       "Calb2              64\n",
       "Ntsr1 Neg          55\n",
       "Rorb               54\n",
       "Nr5a1              48\n",
       "Slc17a6            47\n",
       "PvalbD-Slc32a1     44\n",
       "Scnn1a-Tg3 Neg     23\n",
       "Rbp4 Neg           23\n",
       "Ctgf               23\n",
       "Nkx2-1             22\n",
       "Scnn1a-Tg2         20\n",
       "Ndnf               19\n",
       "Scnn1a-Tg2 Neg     18\n",
       "Cux2 Neg           15\n",
       "Ndnf Neg           12\n",
       "Tac2                8\n",
       "Chat                6\n",
       "Name: cre, dtype: int64"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cell_mtd['cre'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#</th>\n",
       "      <th>Driver Line</th>\n",
       "      <th>Abbreviation</th>\n",
       "      <th>Originating Lab (Donating Investigator)</th>\n",
       "      <th>Primary Reference</th>\n",
       "      <th>Generation Method</th>\n",
       "      <th>Generation Method, more detail</th>\n",
       "      <th>Public Repository</th>\n",
       "      <th>Public Repository Stock #</th>\n",
       "      <th>Repository Strain Name</th>\n",
       "      <th>Data available through the Allen Institute Transgenic Portal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8.0</td>\n",
       "      <td>Ndnf-IRES2-dgCre</td>\n",
       "      <td>Ndnf</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>This paper</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>IRES2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     #       Driver Line Abbreviation Originating Lab (Donating Investigator)  \\\n",
       "7  8.0  Ndnf-IRES2-dgCre         Ndnf       Allen Institute for Brain Science   \n",
       "\n",
       "  Primary Reference Generation Method Generation Method, more detail  \\\n",
       "7        This paper          Knock-in                          IRES2   \n",
       "\n",
       "  Public Repository   Public Repository Stock # Repository Strain Name  \\\n",
       "7                NaN                        NaN                    NaN   \n",
       "\n",
       "  Data available through the Allen Institute Transgenic Portal  \n",
       "7  http://connectivity.brain-map.org/transgenic/s...            "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cre_mtd[df_cre_mtd['Abbreviation'] == 'Ndnf']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "cre_df2 = pd.merge(cre_ref, df_cre_mtd,  how='inner', \n",
    "                   left_on='stock_number',\n",
    "                   right_on = 'Public Repository Stock #')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cre_df2 = pd.merge(df_cre_mtd, cre_ref,  how='inner', \n",
    "                   left_on='Public Repository Stock #',\n",
    "                   right_on = 'stock_number')\n",
    "cre_df2.drop_duplicates(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>originating_lab</th>\n",
       "      <th>stock_number</th>\n",
       "      <th>sub_image_annotation_id</th>\n",
       "      <th>transgenic_line_source_name</th>\n",
       "      <th>transgenic_line_type_code</th>\n",
       "      <th>transgenic_line_type_name</th>\n",
       "      <th>url_prefix</th>\n",
       "      <th>...</th>\n",
       "      <th>Driver Line</th>\n",
       "      <th>Abbreviation</th>\n",
       "      <th>Originating Lab (Donating Investigator)</th>\n",
       "      <th>Primary Reference</th>\n",
       "      <th>Generation Method</th>\n",
       "      <th>Generation Method, more detail</th>\n",
       "      <th>Public Repository</th>\n",
       "      <th>Public Repository Stock #</th>\n",
       "      <th>Repository Strain Name</th>\n",
       "      <th>Data available through the Allen Institute Transgenic Portal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cre expression is enriched in layers 2/3, and ...</td>\n",
       "      <td>306000233</td>\n",
       "      <td>Plxnd1-CreER</td>\n",
       "      <td>Z. Josh Huang</td>\n",
       "      <td>NaN</td>\n",
       "      <td>543091192.0</td>\n",
       "      <td>Other</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Ctgf-2A-dgCre</td>\n",
       "      <td>Ctgf</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>This paper</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>T2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cre expression is enriched in layers 2/3, and ...</td>\n",
       "      <td>306000233</td>\n",
       "      <td>Plxnd1-CreER</td>\n",
       "      <td>Z. Josh Huang</td>\n",
       "      <td>NaN</td>\n",
       "      <td>543091192.0</td>\n",
       "      <td>Other</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Ndnf-IRES2-dgCre</td>\n",
       "      <td>Ndnf</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>This paper</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>IRES2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cre expression is enriched in layers 2/3, and ...</td>\n",
       "      <td>306000233</td>\n",
       "      <td>Plxnd1-CreER</td>\n",
       "      <td>Z. Josh Huang</td>\n",
       "      <td>NaN</td>\n",
       "      <td>543091192.0</td>\n",
       "      <td>Other</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Cre expression is enriched in layers 2/3, and ...</td>\n",
       "      <td>306000233</td>\n",
       "      <td>Plxnd1-CreER</td>\n",
       "      <td>Z. Josh Huang</td>\n",
       "      <td>NaN</td>\n",
       "      <td>543091192.0</td>\n",
       "      <td>Other</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>*Only used for isolation of cortical RNA for o...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Scattered expression in cortical and subcortic...</td>\n",
       "      <td>617166869</td>\n",
       "      <td>Tacr1-T2A-Cre-neo</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>NaN</td>\n",
       "      <td>623914828.0</td>\n",
       "      <td>AIBS</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Ctgf-2A-dgCre</td>\n",
       "      <td>Ctgf</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>This paper</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>T2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Scattered expression in cortical and subcortic...</td>\n",
       "      <td>617166869</td>\n",
       "      <td>Tacr1-T2A-Cre-neo</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>NaN</td>\n",
       "      <td>623914828.0</td>\n",
       "      <td>AIBS</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Ndnf-IRES2-dgCre</td>\n",
       "      <td>Ndnf</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>This paper</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>IRES2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Scattered expression in cortical and subcortic...</td>\n",
       "      <td>617166869</td>\n",
       "      <td>Tacr1-T2A-Cre-neo</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>NaN</td>\n",
       "      <td>623914828.0</td>\n",
       "      <td>AIBS</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Scattered expression in cortical and subcortic...</td>\n",
       "      <td>617166869</td>\n",
       "      <td>Tacr1-T2A-Cre-neo</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>NaN</td>\n",
       "      <td>623914828.0</td>\n",
       "      <td>AIBS</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>*Only used for isolation of cortical RNA for o...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>None</td>\n",
       "      <td>561404352</td>\n",
       "      <td>PhiC31-neo</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Ctgf-2A-dgCre</td>\n",
       "      <td>Ctgf</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>This paper</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>T2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>None</td>\n",
       "      <td>561404352</td>\n",
       "      <td>PhiC31-neo</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Ndnf-IRES2-dgCre</td>\n",
       "      <td>Ndnf</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>This paper</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>IRES2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>None</td>\n",
       "      <td>561404352</td>\n",
       "      <td>PhiC31-neo</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>None</td>\n",
       "      <td>561404352</td>\n",
       "      <td>PhiC31-neo</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>*Only used for isolation of cortical RNA for o...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Ai93 with hygromycin selection cassette.</td>\n",
       "      <td>562981739</td>\n",
       "      <td>Ai93(TITL-GCaMP6f)-hyg</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AIBS</td>\n",
       "      <td>R</td>\n",
       "      <td>reporter</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Ctgf-2A-dgCre</td>\n",
       "      <td>Ctgf</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>This paper</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>T2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Ai93 with hygromycin selection cassette.</td>\n",
       "      <td>562981739</td>\n",
       "      <td>Ai93(TITL-GCaMP6f)-hyg</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AIBS</td>\n",
       "      <td>R</td>\n",
       "      <td>reporter</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Ndnf-IRES2-dgCre</td>\n",
       "      <td>Ndnf</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>This paper</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>IRES2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Ai93 with hygromycin selection cassette.</td>\n",
       "      <td>562981739</td>\n",
       "      <td>Ai93(TITL-GCaMP6f)-hyg</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AIBS</td>\n",
       "      <td>R</td>\n",
       "      <td>reporter</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Ai93 with hygromycin selection cassette.</td>\n",
       "      <td>562981739</td>\n",
       "      <td>Ai93(TITL-GCaMP6f)-hyg</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AIBS</td>\n",
       "      <td>R</td>\n",
       "      <td>reporter</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>*Only used for isolation of cortical RNA for o...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Reporter expression is throughout the cortex, ...</td>\n",
       "      <td>177837979</td>\n",
       "      <td>Drd1a-Cre_EY262</td>\n",
       "      <td>Richard Palmiter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>21593.0</td>\n",
       "      <td>Other</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Ctgf-2A-dgCre</td>\n",
       "      <td>Ctgf</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>This paper</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>T2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Reporter expression is throughout the cortex, ...</td>\n",
       "      <td>177837979</td>\n",
       "      <td>Drd1a-Cre_EY262</td>\n",
       "      <td>Richard Palmiter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>21593.0</td>\n",
       "      <td>Other</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Ndnf-IRES2-dgCre</td>\n",
       "      <td>Ndnf</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>This paper</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>IRES2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Reporter expression is throughout the cortex, ...</td>\n",
       "      <td>177837979</td>\n",
       "      <td>Drd1a-Cre_EY262</td>\n",
       "      <td>Richard Palmiter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>21593.0</td>\n",
       "      <td>Other</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Reporter expression is throughout the cortex, ...</td>\n",
       "      <td>177837979</td>\n",
       "      <td>Drd1a-Cre_EY262</td>\n",
       "      <td>Richard Palmiter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>21593.0</td>\n",
       "      <td>Other</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>*Only used for isolation of cortical RNA for o...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Expressed in subplate, amygdala and hypothalamus.</td>\n",
       "      <td>177836059</td>\n",
       "      <td>EE921-lacZ-CreERT2-Tg3</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>NaN</td>\n",
       "      <td>287070173.0</td>\n",
       "      <td>AIBS</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Ctgf-2A-dgCre</td>\n",
       "      <td>Ctgf</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>This paper</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>T2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Expressed in subplate, amygdala and hypothalamus.</td>\n",
       "      <td>177836059</td>\n",
       "      <td>EE921-lacZ-CreERT2-Tg3</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>NaN</td>\n",
       "      <td>287070173.0</td>\n",
       "      <td>AIBS</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Ndnf-IRES2-dgCre</td>\n",
       "      <td>Ndnf</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>This paper</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>IRES2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Expressed in subplate, amygdala and hypothalamus.</td>\n",
       "      <td>177836059</td>\n",
       "      <td>EE921-lacZ-CreERT2-Tg3</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>NaN</td>\n",
       "      <td>287070173.0</td>\n",
       "      <td>AIBS</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Expressed in subplate, amygdala and hypothalamus.</td>\n",
       "      <td>177836059</td>\n",
       "      <td>EE921-lacZ-CreERT2-Tg3</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>NaN</td>\n",
       "      <td>287070173.0</td>\n",
       "      <td>AIBS</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>*Only used for isolation of cortical RNA for o...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Expression is enriched in specific nuclei of t...</td>\n",
       "      <td>177837364</td>\n",
       "      <td>Kiss1-Cre</td>\n",
       "      <td>Carol Elias</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22032.0</td>\n",
       "      <td>Other</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Ctgf-2A-dgCre</td>\n",
       "      <td>Ctgf</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>This paper</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>T2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Expression is enriched in specific nuclei of t...</td>\n",
       "      <td>177837364</td>\n",
       "      <td>Kiss1-Cre</td>\n",
       "      <td>Carol Elias</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22032.0</td>\n",
       "      <td>Other</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Ndnf-IRES2-dgCre</td>\n",
       "      <td>Ndnf</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>This paper</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>IRES2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Expression is enriched in specific nuclei of t...</td>\n",
       "      <td>177837364</td>\n",
       "      <td>Kiss1-Cre</td>\n",
       "      <td>Carol Elias</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22032.0</td>\n",
       "      <td>Other</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Expression is enriched in specific nuclei of t...</td>\n",
       "      <td>177837364</td>\n",
       "      <td>Kiss1-Cre</td>\n",
       "      <td>Carol Elias</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22032.0</td>\n",
       "      <td>Other</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>*Only used for isolation of cortical RNA for o...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Restricted expression in ventral cortex, thala...</td>\n",
       "      <td>177838679</td>\n",
       "      <td>EE914-lacZ-CreERT2-Tg2</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>NaN</td>\n",
       "      <td>287068503.0</td>\n",
       "      <td>AIBS</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Ctgf-2A-dgCre</td>\n",
       "      <td>Ctgf</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>This paper</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>T2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Restricted expression in ventral cortex, thala...</td>\n",
       "      <td>177838679</td>\n",
       "      <td>EE914-lacZ-CreERT2-Tg2</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>NaN</td>\n",
       "      <td>287068503.0</td>\n",
       "      <td>AIBS</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Ndnf-IRES2-dgCre</td>\n",
       "      <td>Ndnf</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>This paper</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>IRES2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251</th>\n",
       "      <td>Restricted expression in subplate, ventral cor...</td>\n",
       "      <td>177838675</td>\n",
       "      <td>EE921-lacZ-CreERT2-Tg2</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>NaN</td>\n",
       "      <td>287070052.0</td>\n",
       "      <td>AIBS</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>*Only used for isolation of cortical RNA for o...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>Cre expression is restricted to specific popul...</td>\n",
       "      <td>177836019</td>\n",
       "      <td>Esr1-2A-Cre</td>\n",
       "      <td>David Anderson</td>\n",
       "      <td>NaN</td>\n",
       "      <td>298737224.0</td>\n",
       "      <td>Other</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Ctgf-2A-dgCre</td>\n",
       "      <td>Ctgf</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>This paper</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>T2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253</th>\n",
       "      <td>Cre expression is restricted to specific popul...</td>\n",
       "      <td>177836019</td>\n",
       "      <td>Esr1-2A-Cre</td>\n",
       "      <td>David Anderson</td>\n",
       "      <td>NaN</td>\n",
       "      <td>298737224.0</td>\n",
       "      <td>Other</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Ndnf-IRES2-dgCre</td>\n",
       "      <td>Ndnf</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>This paper</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>IRES2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>Cre expression is restricted to specific popul...</td>\n",
       "      <td>177836019</td>\n",
       "      <td>Esr1-2A-Cre</td>\n",
       "      <td>David Anderson</td>\n",
       "      <td>NaN</td>\n",
       "      <td>298737224.0</td>\n",
       "      <td>Other</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>Cre expression is restricted to specific popul...</td>\n",
       "      <td>177836019</td>\n",
       "      <td>Esr1-2A-Cre</td>\n",
       "      <td>David Anderson</td>\n",
       "      <td>NaN</td>\n",
       "      <td>298737224.0</td>\n",
       "      <td>Other</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>*Only used for isolation of cortical RNA for o...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256</th>\n",
       "      <td>Enrichment in cortical layer 5 and layer 6 neu...</td>\n",
       "      <td>617170129</td>\n",
       "      <td>Npr3-IRES2-Cre-neo</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>NaN</td>\n",
       "      <td>623891829.0</td>\n",
       "      <td>AIBS</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Ctgf-2A-dgCre</td>\n",
       "      <td>Ctgf</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>This paper</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>T2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>Enrichment in cortical layer 5 and layer 6 neu...</td>\n",
       "      <td>617170129</td>\n",
       "      <td>Npr3-IRES2-Cre-neo</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>NaN</td>\n",
       "      <td>623891829.0</td>\n",
       "      <td>AIBS</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Ndnf-IRES2-dgCre</td>\n",
       "      <td>Ndnf</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>This paper</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>IRES2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>Enrichment in cortical layer 5 and layer 6 neu...</td>\n",
       "      <td>617170129</td>\n",
       "      <td>Npr3-IRES2-Cre-neo</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>NaN</td>\n",
       "      <td>623891829.0</td>\n",
       "      <td>AIBS</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>Enrichment in cortical layer 5 and layer 6 neu...</td>\n",
       "      <td>617170129</td>\n",
       "      <td>Npr3-IRES2-Cre-neo</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>NaN</td>\n",
       "      <td>623891829.0</td>\n",
       "      <td>AIBS</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>*Only used for isolation of cortical RNA for o...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>Specific to GABAergic neurons. Enriched in the...</td>\n",
       "      <td>177838584</td>\n",
       "      <td>Gad2-IRES-Cre</td>\n",
       "      <td>Z. Josh Huang</td>\n",
       "      <td>10802.0</td>\n",
       "      <td>21959.0</td>\n",
       "      <td>JAX</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>http://jaxmice.jax.org/strain/</td>\n",
       "      <td>...</td>\n",
       "      <td>Gad2-IRES-Cre</td>\n",
       "      <td>Gad2</td>\n",
       "      <td>Z. Josh Huang</td>\n",
       "      <td>Taniguchi et al., Neuron 2011</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>IRES</td>\n",
       "      <td>The Jackson Laboratory</td>\n",
       "      <td>10802.0</td>\n",
       "      <td>STOCK Gad2tm2(cre)Zjh/J</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>Expressed in cholinergic neurons - restricted ...</td>\n",
       "      <td>553502711</td>\n",
       "      <td>Chat-IRES-Cre-neo</td>\n",
       "      <td>Bradford Lowell</td>\n",
       "      <td>6410.0</td>\n",
       "      <td>21944.0</td>\n",
       "      <td>JAX</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>http://jaxmice.jax.org/strain/</td>\n",
       "      <td>...</td>\n",
       "      <td>Chat-IRES-Cre</td>\n",
       "      <td>Chat</td>\n",
       "      <td>Brad Lowell</td>\n",
       "      <td>Rossi et al., Cell Metab. 2011</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>IRES</td>\n",
       "      <td>The Jackson Laboratory</td>\n",
       "      <td>6410.0</td>\n",
       "      <td>B6;129S6-Chattm1(cre)Lowl/J</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>Reporter expression is detected in a subset of...</td>\n",
       "      <td>298341195</td>\n",
       "      <td>Htr3a-Cre_NO152</td>\n",
       "      <td>Nathaniel Heintz and Charles Gerfen</td>\n",
       "      <td>36680.0</td>\n",
       "      <td>308565632.0</td>\n",
       "      <td>MMRRC</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>http://www.mmrrc.org/catalog/getSDS.jsp?mmrrc_id=</td>\n",
       "      <td>...</td>\n",
       "      <td>Htr3a-Cre_NO152</td>\n",
       "      <td>Htr3a</td>\n",
       "      <td>Nathaniel Heintz and Charles Gerfen</td>\n",
       "      <td>Gerfen et al., Neuron 2013</td>\n",
       "      <td>Transgenic</td>\n",
       "      <td>BAC</td>\n",
       "      <td>MMRRC</td>\n",
       "      <td>36680.0</td>\n",
       "      <td>STOCK Tg(Htr3a-cre)NO152Gsat/Mmucd</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>Expressed in restricted and/or sparse populati...</td>\n",
       "      <td>177837858</td>\n",
       "      <td>Pvalb-IRES-Cre</td>\n",
       "      <td>Silvia Arber</td>\n",
       "      <td>8069.0</td>\n",
       "      <td>22037.0</td>\n",
       "      <td>JAX</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>http://jaxmice.jax.org/strain/</td>\n",
       "      <td>...</td>\n",
       "      <td>Pvalb-IRES-Cre</td>\n",
       "      <td>Pvalb</td>\n",
       "      <td>Silvia Arber</td>\n",
       "      <td>Hippenmeyer et al., PLoS Biol 2005</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>IRES</td>\n",
       "      <td>The Jackson Laboratory</td>\n",
       "      <td>8069.0</td>\n",
       "      <td>B6;129P2-Pvalbtm1(cre)Arbr/J</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>264</th>\n",
       "      <td>Expressed in restricted populations within the...</td>\n",
       "      <td>177839331</td>\n",
       "      <td>Nr5a1-Cre</td>\n",
       "      <td>Bradford Lowell</td>\n",
       "      <td>6364.0</td>\n",
       "      <td>21972.0</td>\n",
       "      <td>JAX</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>http://jaxmice.jax.org/strain/</td>\n",
       "      <td>...</td>\n",
       "      <td>Nr5a1-Cre</td>\n",
       "      <td>Nr5a1</td>\n",
       "      <td>Brad Lowell</td>\n",
       "      <td>Dhillon et al., Neuron 2006</td>\n",
       "      <td>Transgenic</td>\n",
       "      <td>BAC</td>\n",
       "      <td>The Jackson Laboratory</td>\n",
       "      <td>6364.0</td>\n",
       "      <td>FVB-Tg(Nr5a1-cre)2Lowl/J</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>Strong scattered expression throughout the bra...</td>\n",
       "      <td>177838710</td>\n",
       "      <td>Sst-IRES-Cre</td>\n",
       "      <td>Z. Josh Huang</td>\n",
       "      <td>13044.0</td>\n",
       "      <td>157876407.0</td>\n",
       "      <td>JAX</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>http://jaxmice.jax.org/strain/</td>\n",
       "      <td>...</td>\n",
       "      <td>Sst-IRES-Cre</td>\n",
       "      <td>Sst</td>\n",
       "      <td>Z. Josh Huang</td>\n",
       "      <td>Taniguchi et al., Neuron 2011</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>IRES</td>\n",
       "      <td>The Jackson Laboratory</td>\n",
       "      <td>13044.0</td>\n",
       "      <td>STOCK Ssttm2.1(cre)Zjh/J</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>Scattered expression throughout the cortex. En...</td>\n",
       "      <td>182103856</td>\n",
       "      <td>Pvalb-T2A-FlpO</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>22730.0</td>\n",
       "      <td>550921303.0</td>\n",
       "      <td>JAX</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>http://jaxmice.jax.org/strain/</td>\n",
       "      <td>...</td>\n",
       "      <td>Pvalb-2A-Flpo</td>\n",
       "      <td>PvalbF</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>Madisen et al., Neuron 2015</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>T2A</td>\n",
       "      <td>The Jackson Laboratory</td>\n",
       "      <td>22730.0</td>\n",
       "      <td>B6.Cg-Pvalbtm4.1(FLPo)Hze/J</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267</th>\n",
       "      <td>Enriched in cortical layer 4 and in restricted...</td>\n",
       "      <td>177837516</td>\n",
       "      <td>Scnn1a-Tg3-Cre</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>9613.0</td>\n",
       "      <td>22040.0</td>\n",
       "      <td>JAX</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>http://jaxmice.jax.org/strain/</td>\n",
       "      <td>...</td>\n",
       "      <td>Scnn1a-Tg3-Cre</td>\n",
       "      <td>Scnn1a-Tg3</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>Madisen et al., Nat. Neu. 2010</td>\n",
       "      <td>Transgenic</td>\n",
       "      <td>BAC</td>\n",
       "      <td>The Jackson Laboratory</td>\n",
       "      <td>9613.0</td>\n",
       "      <td>B6;C3-Tg(Scnn1a-cre)3Aibs/J</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>268</th>\n",
       "      <td>Strong scattered cells in restricted populatio...</td>\n",
       "      <td>177835613</td>\n",
       "      <td>Nkx2-1-CreERT2</td>\n",
       "      <td>Z. Josh Huang</td>\n",
       "      <td>14552.0</td>\n",
       "      <td>178397462.0</td>\n",
       "      <td>JAX</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>http://jaxmice.jax.org/strain/</td>\n",
       "      <td>...</td>\n",
       "      <td>Nkx2.1-CreERT2</td>\n",
       "      <td>Nkx2.1</td>\n",
       "      <td>Z. Josh Huang</td>\n",
       "      <td>Taniguchi et al., Science 2013</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>Direct</td>\n",
       "      <td>The Jackson Laboratory</td>\n",
       "      <td>14552.0</td>\n",
       "      <td>STOCK Nkx2-1tm1.1(cre/ERT2)Zjh/J</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269</th>\n",
       "      <td>Strong scattered expression throughout the bra...</td>\n",
       "      <td>177838622</td>\n",
       "      <td>Vip-IRES-Cre</td>\n",
       "      <td>Z. Josh Huang</td>\n",
       "      <td>10908.0</td>\n",
       "      <td>157876322.0</td>\n",
       "      <td>JAX</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>http://jaxmice.jax.org/strain/</td>\n",
       "      <td>...</td>\n",
       "      <td>Vip-IRES-Cre</td>\n",
       "      <td>Vip</td>\n",
       "      <td>Z. Josh Huang</td>\n",
       "      <td>Taniguchi et al., Neuron 2011</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>IRES</td>\n",
       "      <td>The Jackson Laboratory</td>\n",
       "      <td>10908.0</td>\n",
       "      <td>STOCK Viptm1(cre)Zjh/J</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270</th>\n",
       "      <td>Scattered expression in caudate, septum, hypot...</td>\n",
       "      <td>177838715</td>\n",
       "      <td>Tac1-IRES2-Cre</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>21877.0</td>\n",
       "      <td>158462163.0</td>\n",
       "      <td>JAX</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>http://jaxmice.jax.org/strain/</td>\n",
       "      <td>...</td>\n",
       "      <td>Tac1-IRES2-Cre</td>\n",
       "      <td>Tac1</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>Harris et al., Front. Neural Circuits, 2014</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>IRES2</td>\n",
       "      <td>The Jackson Laboratory</td>\n",
       "      <td>21877.0</td>\n",
       "      <td>B6;129S-Tac1tm1.1(cre)Hze/J</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>271</th>\n",
       "      <td>Enriched in restricted populations within the ...</td>\n",
       "      <td>177837281</td>\n",
       "      <td>Calb2-IRES-Cre</td>\n",
       "      <td>Z. Josh Huang</td>\n",
       "      <td>10774.0</td>\n",
       "      <td>21940.0</td>\n",
       "      <td>JAX</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>http://jaxmice.jax.org/strain/</td>\n",
       "      <td>...</td>\n",
       "      <td>Calb2-IRES-Cre</td>\n",
       "      <td>Calb2</td>\n",
       "      <td>Z. Josh Huang</td>\n",
       "      <td>Taniguchi et al., Neuron 2011</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>IRES</td>\n",
       "      <td>The Jackson Laboratory</td>\n",
       "      <td>10774.0</td>\n",
       "      <td>B6(Cg)-Calb2tm1(cre)Zjh/J</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272</th>\n",
       "      <td>Enriched in superficial layer 5 neurons of cor...</td>\n",
       "      <td>177838633</td>\n",
       "      <td>Trib2-F2A-CreERT2</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>22865.0</td>\n",
       "      <td>22054.0</td>\n",
       "      <td>JAX</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>http://jaxmice.jax.org/strain/</td>\n",
       "      <td>...</td>\n",
       "      <td>Trib2-2A-CreERT2</td>\n",
       "      <td>Trib2*</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>Harris et al., Front. Neural Circuits, 2014</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>F2A</td>\n",
       "      <td>The Jackson Laboratory</td>\n",
       "      <td>22865.0</td>\n",
       "      <td>B6.Cg-Trib2tm1.1(cre/ERT2)Hze/J</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>273</th>\n",
       "      <td>Enriched in cortical layers 2/3/4, thalamus, m...</td>\n",
       "      <td>177839004</td>\n",
       "      <td>Cux2-CreERT2</td>\n",
       "      <td>Ulrich Mueller</td>\n",
       "      <td>32779.0</td>\n",
       "      <td>21882.0</td>\n",
       "      <td>MMRRC</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>http://www.mmrrc.org/catalog/getSDS.jsp?mmrrc_id=</td>\n",
       "      <td>...</td>\n",
       "      <td>Cux2-CreERT2</td>\n",
       "      <td>Cux2</td>\n",
       "      <td>Ulrich Mueller</td>\n",
       "      <td>Franco et al., Science 2012</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>Direct</td>\n",
       "      <td>MMRRC</td>\n",
       "      <td>32779.0</td>\n",
       "      <td>B6(Cg)-Cux2&lt;tm3.1(cre/ERT2)Mull&gt;/Mmmh</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274</th>\n",
       "      <td>Reporter expression in sparse and/or restricte...</td>\n",
       "      <td>177837481</td>\n",
       "      <td>Scnn1a-Tg2-Cre</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>9112.0</td>\n",
       "      <td>22039.0</td>\n",
       "      <td>JAX</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>http://jaxmice.jax.org/strain/</td>\n",
       "      <td>...</td>\n",
       "      <td>Scnn1a-Tg2-Cre</td>\n",
       "      <td>Scnn1a-Tg2</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>Madisen et al., Nat. Neu. 2010</td>\n",
       "      <td>Transgenic</td>\n",
       "      <td>BAC</td>\n",
       "      <td>The Jackson Laboratory</td>\n",
       "      <td>9112.0</td>\n",
       "      <td>B6;C3-Tg(Scnn1a-cre)2Aibs/J</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>275</th>\n",
       "      <td>Dre expression is scattered throughout many re...</td>\n",
       "      <td>515746198</td>\n",
       "      <td>Pvalb-T2A-Dre</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>21190.0</td>\n",
       "      <td>471756374.0</td>\n",
       "      <td>JAX</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>http://jaxmice.jax.org/strain/</td>\n",
       "      <td>...</td>\n",
       "      <td>Pvalb-2A-Dre</td>\n",
       "      <td>PvalbD</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>Madisen et al., Neuron 2015</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>T2A</td>\n",
       "      <td>The Jackson Laboratory</td>\n",
       "      <td>21190.0</td>\n",
       "      <td>B6.Cg-Pvalbtm3.1(dreo)Hze/J</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>Specific to cortical layer 6 neurons.</td>\n",
       "      <td>177837713</td>\n",
       "      <td>Ntsr1-Cre_GN220</td>\n",
       "      <td>Nathaniel Heintz and Charles Gerfen</td>\n",
       "      <td>30648.0</td>\n",
       "      <td>21592.0</td>\n",
       "      <td>MMRRC</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>http://www.mmrrc.org/catalog/getSDS.jsp?mmrrc_id=</td>\n",
       "      <td>...</td>\n",
       "      <td>Ntsr1-Cre_GN220</td>\n",
       "      <td>Ntsr1</td>\n",
       "      <td>Nathaniel Heintz and Charles Gerfen</td>\n",
       "      <td>Gerfen et al., Neuron 2013</td>\n",
       "      <td>Transgenic</td>\n",
       "      <td>BAC</td>\n",
       "      <td>MMRRC</td>\n",
       "      <td>30648.0</td>\n",
       "      <td>B6.FVB(Cg)-Tg(Ntsr1-cre)GN220Gsat/Mmucd</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>Enriched in cortical layer 5 and the dentate g...</td>\n",
       "      <td>177838435</td>\n",
       "      <td>Rbp4-Cre_KL100</td>\n",
       "      <td>Nathaniel Heintz and Charles Gerfen</td>\n",
       "      <td>31125.0</td>\n",
       "      <td>120818676.0</td>\n",
       "      <td>MMRRC</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>http://www.mmrrc.org/catalog/getSDS.jsp?mmrrc_id=</td>\n",
       "      <td>...</td>\n",
       "      <td>Rbp4-Cre_KL100</td>\n",
       "      <td>Rbp4</td>\n",
       "      <td>Nathaniel Heintz and Charles Gerfen</td>\n",
       "      <td>Gerfen et al., Neuron 2013</td>\n",
       "      <td>Transgenic</td>\n",
       "      <td>BAC</td>\n",
       "      <td>MMRRC</td>\n",
       "      <td>31125.0</td>\n",
       "      <td>STOCK Tg(Rbp4-cre)KL100Gsat/Mmucd</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278</th>\n",
       "      <td>Enriched in cortical layer 5. Also expressed i...</td>\n",
       "      <td>182693192</td>\n",
       "      <td>Chrna2-Cre_OE25</td>\n",
       "      <td>Nathaniel Heintz and Charles Gerfen</td>\n",
       "      <td>36502.0</td>\n",
       "      <td>286888096.0</td>\n",
       "      <td>MMRRC</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>http://www.mmrrc.org/catalog/getSDS.jsp?mmrrc_id=</td>\n",
       "      <td>...</td>\n",
       "      <td>Chrna2-Cre_OE25</td>\n",
       "      <td>Chrna2</td>\n",
       "      <td>Nathaniel Heintz and Charles Gerfen</td>\n",
       "      <td>Gerfen et al., Neuron 2013</td>\n",
       "      <td>Transgenic</td>\n",
       "      <td>BAC</td>\n",
       "      <td>MMRRC</td>\n",
       "      <td>36502.0</td>\n",
       "      <td>STOCK Tg(Chrna2-cre)OE25Gsat/Mmucd</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>Expressed in the molecular and granular layer ...</td>\n",
       "      <td>177839406</td>\n",
       "      <td>Nos1-CreERT2</td>\n",
       "      <td>Z. Josh Huang</td>\n",
       "      <td>14541.0</td>\n",
       "      <td>21894.0</td>\n",
       "      <td>JAX</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>http://jaxmice.jax.org/strain/</td>\n",
       "      <td>...</td>\n",
       "      <td>Nos1-CreERT2</td>\n",
       "      <td>Nos1</td>\n",
       "      <td>Z. Josh Huang</td>\n",
       "      <td>Taniguchi et al., Neuron 2011</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>Direct</td>\n",
       "      <td>The Jackson Laboratory</td>\n",
       "      <td>14541.0</td>\n",
       "      <td>B6;129S-Nos1tm1.1(cre/ERT2)Zjh/J</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280</th>\n",
       "      <td>Strong expression in the zonal layer of the su...</td>\n",
       "      <td>179697864</td>\n",
       "      <td>Rorb-IRES2-Cre</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>23526.0</td>\n",
       "      <td>263635668.0</td>\n",
       "      <td>JAX</td>\n",
       "      <td>D</td>\n",
       "      <td>driver</td>\n",
       "      <td>http://jaxmice.jax.org/strain/</td>\n",
       "      <td>...</td>\n",
       "      <td>Rorb-IRES2-Cre</td>\n",
       "      <td>Rorb</td>\n",
       "      <td>Allen Institute for Brain Science</td>\n",
       "      <td>Harris et al., Front. Neural Circuits, 2014</td>\n",
       "      <td>Knock-in</td>\n",
       "      <td>IRES2</td>\n",
       "      <td>The Jackson Laboratory</td>\n",
       "      <td>23526.0</td>\n",
       "      <td>B6;129S-Rorbtm1.1(cre)Hze/J</td>\n",
       "      <td>http://connectivity.brain-map.org/transgenic/s...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>281 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           description         id  \\\n",
       "0    Cre expression is enriched in layers 2/3, and ...  306000233   \n",
       "1    Cre expression is enriched in layers 2/3, and ...  306000233   \n",
       "2    Cre expression is enriched in layers 2/3, and ...  306000233   \n",
       "3    Cre expression is enriched in layers 2/3, and ...  306000233   \n",
       "4    Scattered expression in cortical and subcortic...  617166869   \n",
       "5    Scattered expression in cortical and subcortic...  617166869   \n",
       "6    Scattered expression in cortical and subcortic...  617166869   \n",
       "7    Scattered expression in cortical and subcortic...  617166869   \n",
       "8                                                 None  561404352   \n",
       "9                                                 None  561404352   \n",
       "10                                                None  561404352   \n",
       "11                                                None  561404352   \n",
       "12            Ai93 with hygromycin selection cassette.  562981739   \n",
       "13            Ai93 with hygromycin selection cassette.  562981739   \n",
       "14            Ai93 with hygromycin selection cassette.  562981739   \n",
       "15            Ai93 with hygromycin selection cassette.  562981739   \n",
       "16   Reporter expression is throughout the cortex, ...  177837979   \n",
       "17   Reporter expression is throughout the cortex, ...  177837979   \n",
       "18   Reporter expression is throughout the cortex, ...  177837979   \n",
       "19   Reporter expression is throughout the cortex, ...  177837979   \n",
       "20   Expressed in subplate, amygdala and hypothalamus.  177836059   \n",
       "21   Expressed in subplate, amygdala and hypothalamus.  177836059   \n",
       "22   Expressed in subplate, amygdala and hypothalamus.  177836059   \n",
       "23   Expressed in subplate, amygdala and hypothalamus.  177836059   \n",
       "24   Expression is enriched in specific nuclei of t...  177837364   \n",
       "25   Expression is enriched in specific nuclei of t...  177837364   \n",
       "26   Expression is enriched in specific nuclei of t...  177837364   \n",
       "27   Expression is enriched in specific nuclei of t...  177837364   \n",
       "28   Restricted expression in ventral cortex, thala...  177838679   \n",
       "29   Restricted expression in ventral cortex, thala...  177838679   \n",
       "..                                                 ...        ...   \n",
       "251  Restricted expression in subplate, ventral cor...  177838675   \n",
       "252  Cre expression is restricted to specific popul...  177836019   \n",
       "253  Cre expression is restricted to specific popul...  177836019   \n",
       "254  Cre expression is restricted to specific popul...  177836019   \n",
       "255  Cre expression is restricted to specific popul...  177836019   \n",
       "256  Enrichment in cortical layer 5 and layer 6 neu...  617170129   \n",
       "257  Enrichment in cortical layer 5 and layer 6 neu...  617170129   \n",
       "258  Enrichment in cortical layer 5 and layer 6 neu...  617170129   \n",
       "259  Enrichment in cortical layer 5 and layer 6 neu...  617170129   \n",
       "260  Specific to GABAergic neurons. Enriched in the...  177838584   \n",
       "261  Expressed in cholinergic neurons - restricted ...  553502711   \n",
       "262  Reporter expression is detected in a subset of...  298341195   \n",
       "263  Expressed in restricted and/or sparse populati...  177837858   \n",
       "264  Expressed in restricted populations within the...  177839331   \n",
       "265  Strong scattered expression throughout the bra...  177838710   \n",
       "266  Scattered expression throughout the cortex. En...  182103856   \n",
       "267  Enriched in cortical layer 4 and in restricted...  177837516   \n",
       "268  Strong scattered cells in restricted populatio...  177835613   \n",
       "269  Strong scattered expression throughout the bra...  177838622   \n",
       "270  Scattered expression in caudate, septum, hypot...  177838715   \n",
       "271  Enriched in restricted populations within the ...  177837281   \n",
       "272  Enriched in superficial layer 5 neurons of cor...  177838633   \n",
       "273  Enriched in cortical layers 2/3/4, thalamus, m...  177839004   \n",
       "274  Reporter expression in sparse and/or restricte...  177837481   \n",
       "275  Dre expression is scattered throughout many re...  515746198   \n",
       "276              Specific to cortical layer 6 neurons.  177837713   \n",
       "277  Enriched in cortical layer 5 and the dentate g...  177838435   \n",
       "278  Enriched in cortical layer 5. Also expressed i...  182693192   \n",
       "279  Expressed in the molecular and granular layer ...  177839406   \n",
       "280  Strong expression in the zonal layer of the su...  179697864   \n",
       "\n",
       "                       name                      originating_lab  \\\n",
       "0              Plxnd1-CreER                        Z. Josh Huang   \n",
       "1              Plxnd1-CreER                        Z. Josh Huang   \n",
       "2              Plxnd1-CreER                        Z. Josh Huang   \n",
       "3              Plxnd1-CreER                        Z. Josh Huang   \n",
       "4         Tacr1-T2A-Cre-neo    Allen Institute for Brain Science   \n",
       "5         Tacr1-T2A-Cre-neo    Allen Institute for Brain Science   \n",
       "6         Tacr1-T2A-Cre-neo    Allen Institute for Brain Science   \n",
       "7         Tacr1-T2A-Cre-neo    Allen Institute for Brain Science   \n",
       "8                PhiC31-neo                                 None   \n",
       "9                PhiC31-neo                                 None   \n",
       "10               PhiC31-neo                                 None   \n",
       "11               PhiC31-neo                                 None   \n",
       "12   Ai93(TITL-GCaMP6f)-hyg    Allen Institute for Brain Science   \n",
       "13   Ai93(TITL-GCaMP6f)-hyg    Allen Institute for Brain Science   \n",
       "14   Ai93(TITL-GCaMP6f)-hyg    Allen Institute for Brain Science   \n",
       "15   Ai93(TITL-GCaMP6f)-hyg    Allen Institute for Brain Science   \n",
       "16          Drd1a-Cre_EY262                     Richard Palmiter   \n",
       "17          Drd1a-Cre_EY262                     Richard Palmiter   \n",
       "18          Drd1a-Cre_EY262                     Richard Palmiter   \n",
       "19          Drd1a-Cre_EY262                     Richard Palmiter   \n",
       "20   EE921-lacZ-CreERT2-Tg3    Allen Institute for Brain Science   \n",
       "21   EE921-lacZ-CreERT2-Tg3    Allen Institute for Brain Science   \n",
       "22   EE921-lacZ-CreERT2-Tg3    Allen Institute for Brain Science   \n",
       "23   EE921-lacZ-CreERT2-Tg3    Allen Institute for Brain Science   \n",
       "24                Kiss1-Cre                          Carol Elias   \n",
       "25                Kiss1-Cre                          Carol Elias   \n",
       "26                Kiss1-Cre                          Carol Elias   \n",
       "27                Kiss1-Cre                          Carol Elias   \n",
       "28   EE914-lacZ-CreERT2-Tg2    Allen Institute for Brain Science   \n",
       "29   EE914-lacZ-CreERT2-Tg2    Allen Institute for Brain Science   \n",
       "..                      ...                                  ...   \n",
       "251  EE921-lacZ-CreERT2-Tg2    Allen Institute for Brain Science   \n",
       "252             Esr1-2A-Cre                       David Anderson   \n",
       "253             Esr1-2A-Cre                       David Anderson   \n",
       "254             Esr1-2A-Cre                       David Anderson   \n",
       "255             Esr1-2A-Cre                       David Anderson   \n",
       "256      Npr3-IRES2-Cre-neo    Allen Institute for Brain Science   \n",
       "257      Npr3-IRES2-Cre-neo    Allen Institute for Brain Science   \n",
       "258      Npr3-IRES2-Cre-neo    Allen Institute for Brain Science   \n",
       "259      Npr3-IRES2-Cre-neo    Allen Institute for Brain Science   \n",
       "260           Gad2-IRES-Cre                        Z. Josh Huang   \n",
       "261       Chat-IRES-Cre-neo                      Bradford Lowell   \n",
       "262         Htr3a-Cre_NO152  Nathaniel Heintz and Charles Gerfen   \n",
       "263          Pvalb-IRES-Cre                         Silvia Arber   \n",
       "264               Nr5a1-Cre                      Bradford Lowell   \n",
       "265            Sst-IRES-Cre                        Z. Josh Huang   \n",
       "266          Pvalb-T2A-FlpO    Allen Institute for Brain Science   \n",
       "267          Scnn1a-Tg3-Cre    Allen Institute for Brain Science   \n",
       "268          Nkx2-1-CreERT2                        Z. Josh Huang   \n",
       "269            Vip-IRES-Cre                        Z. Josh Huang   \n",
       "270          Tac1-IRES2-Cre    Allen Institute for Brain Science   \n",
       "271          Calb2-IRES-Cre                        Z. Josh Huang   \n",
       "272       Trib2-F2A-CreERT2    Allen Institute for Brain Science   \n",
       "273            Cux2-CreERT2                       Ulrich Mueller   \n",
       "274          Scnn1a-Tg2-Cre    Allen Institute for Brain Science   \n",
       "275           Pvalb-T2A-Dre    Allen Institute for Brain Science   \n",
       "276         Ntsr1-Cre_GN220  Nathaniel Heintz and Charles Gerfen   \n",
       "277          Rbp4-Cre_KL100  Nathaniel Heintz and Charles Gerfen   \n",
       "278         Chrna2-Cre_OE25  Nathaniel Heintz and Charles Gerfen   \n",
       "279            Nos1-CreERT2                        Z. Josh Huang   \n",
       "280          Rorb-IRES2-Cre    Allen Institute for Brain Science   \n",
       "\n",
       "     stock_number  sub_image_annotation_id transgenic_line_source_name  \\\n",
       "0             NaN              543091192.0                       Other   \n",
       "1             NaN              543091192.0                       Other   \n",
       "2             NaN              543091192.0                       Other   \n",
       "3             NaN              543091192.0                       Other   \n",
       "4             NaN              623914828.0                        AIBS   \n",
       "5             NaN              623914828.0                        AIBS   \n",
       "6             NaN              623914828.0                        AIBS   \n",
       "7             NaN              623914828.0                        AIBS   \n",
       "8             NaN                      NaN                        None   \n",
       "9             NaN                      NaN                        None   \n",
       "10            NaN                      NaN                        None   \n",
       "11            NaN                      NaN                        None   \n",
       "12            NaN                      NaN                        AIBS   \n",
       "13            NaN                      NaN                        AIBS   \n",
       "14            NaN                      NaN                        AIBS   \n",
       "15            NaN                      NaN                        AIBS   \n",
       "16            NaN                  21593.0                       Other   \n",
       "17            NaN                  21593.0                       Other   \n",
       "18            NaN                  21593.0                       Other   \n",
       "19            NaN                  21593.0                       Other   \n",
       "20            NaN              287070173.0                        AIBS   \n",
       "21            NaN              287070173.0                        AIBS   \n",
       "22            NaN              287070173.0                        AIBS   \n",
       "23            NaN              287070173.0                        AIBS   \n",
       "24            NaN                  22032.0                       Other   \n",
       "25            NaN                  22032.0                       Other   \n",
       "26            NaN                  22032.0                       Other   \n",
       "27            NaN                  22032.0                       Other   \n",
       "28            NaN              287068503.0                        AIBS   \n",
       "29            NaN              287068503.0                        AIBS   \n",
       "..            ...                      ...                         ...   \n",
       "251           NaN              287070052.0                        AIBS   \n",
       "252           NaN              298737224.0                       Other   \n",
       "253           NaN              298737224.0                       Other   \n",
       "254           NaN              298737224.0                       Other   \n",
       "255           NaN              298737224.0                       Other   \n",
       "256           NaN              623891829.0                        AIBS   \n",
       "257           NaN              623891829.0                        AIBS   \n",
       "258           NaN              623891829.0                        AIBS   \n",
       "259           NaN              623891829.0                        AIBS   \n",
       "260       10802.0                  21959.0                         JAX   \n",
       "261        6410.0                  21944.0                         JAX   \n",
       "262       36680.0              308565632.0                       MMRRC   \n",
       "263        8069.0                  22037.0                         JAX   \n",
       "264        6364.0                  21972.0                         JAX   \n",
       "265       13044.0              157876407.0                         JAX   \n",
       "266       22730.0              550921303.0                         JAX   \n",
       "267        9613.0                  22040.0                         JAX   \n",
       "268       14552.0              178397462.0                         JAX   \n",
       "269       10908.0              157876322.0                         JAX   \n",
       "270       21877.0              158462163.0                         JAX   \n",
       "271       10774.0                  21940.0                         JAX   \n",
       "272       22865.0                  22054.0                         JAX   \n",
       "273       32779.0                  21882.0                       MMRRC   \n",
       "274        9112.0                  22039.0                         JAX   \n",
       "275       21190.0              471756374.0                         JAX   \n",
       "276       30648.0                  21592.0                       MMRRC   \n",
       "277       31125.0              120818676.0                       MMRRC   \n",
       "278       36502.0              286888096.0                       MMRRC   \n",
       "279       14541.0                  21894.0                         JAX   \n",
       "280       23526.0              263635668.0                         JAX   \n",
       "\n",
       "    transgenic_line_type_code transgenic_line_type_name  \\\n",
       "0                           D                    driver   \n",
       "1                           D                    driver   \n",
       "2                           D                    driver   \n",
       "3                           D                    driver   \n",
       "4                           D                    driver   \n",
       "5                           D                    driver   \n",
       "6                           D                    driver   \n",
       "7                           D                    driver   \n",
       "8                           D                    driver   \n",
       "9                           D                    driver   \n",
       "10                          D                    driver   \n",
       "11                          D                    driver   \n",
       "12                          R                  reporter   \n",
       "13                          R                  reporter   \n",
       "14                          R                  reporter   \n",
       "15                          R                  reporter   \n",
       "16                          D                    driver   \n",
       "17                          D                    driver   \n",
       "18                          D                    driver   \n",
       "19                          D                    driver   \n",
       "20                          D                    driver   \n",
       "21                          D                    driver   \n",
       "22                          D                    driver   \n",
       "23                          D                    driver   \n",
       "24                          D                    driver   \n",
       "25                          D                    driver   \n",
       "26                          D                    driver   \n",
       "27                          D                    driver   \n",
       "28                          D                    driver   \n",
       "29                          D                    driver   \n",
       "..                        ...                       ...   \n",
       "251                         D                    driver   \n",
       "252                         D                    driver   \n",
       "253                         D                    driver   \n",
       "254                         D                    driver   \n",
       "255                         D                    driver   \n",
       "256                         D                    driver   \n",
       "257                         D                    driver   \n",
       "258                         D                    driver   \n",
       "259                         D                    driver   \n",
       "260                         D                    driver   \n",
       "261                         D                    driver   \n",
       "262                         D                    driver   \n",
       "263                         D                    driver   \n",
       "264                         D                    driver   \n",
       "265                         D                    driver   \n",
       "266                         D                    driver   \n",
       "267                         D                    driver   \n",
       "268                         D                    driver   \n",
       "269                         D                    driver   \n",
       "270                         D                    driver   \n",
       "271                         D                    driver   \n",
       "272                         D                    driver   \n",
       "273                         D                    driver   \n",
       "274                         D                    driver   \n",
       "275                         D                    driver   \n",
       "276                         D                    driver   \n",
       "277                         D                    driver   \n",
       "278                         D                    driver   \n",
       "279                         D                    driver   \n",
       "280                         D                    driver   \n",
       "\n",
       "                                            url_prefix  ...  \\\n",
       "0                                                 None  ...   \n",
       "1                                                 None  ...   \n",
       "2                                                 None  ...   \n",
       "3                                                 None  ...   \n",
       "4                                                 None  ...   \n",
       "5                                                 None  ...   \n",
       "6                                                 None  ...   \n",
       "7                                                 None  ...   \n",
       "8                                                 None  ...   \n",
       "9                                                 None  ...   \n",
       "10                                                None  ...   \n",
       "11                                                None  ...   \n",
       "12                                                None  ...   \n",
       "13                                                None  ...   \n",
       "14                                                None  ...   \n",
       "15                                                None  ...   \n",
       "16                                                None  ...   \n",
       "17                                                None  ...   \n",
       "18                                                None  ...   \n",
       "19                                                None  ...   \n",
       "20                                                None  ...   \n",
       "21                                                None  ...   \n",
       "22                                                None  ...   \n",
       "23                                                None  ...   \n",
       "24                                                None  ...   \n",
       "25                                                None  ...   \n",
       "26                                                None  ...   \n",
       "27                                                None  ...   \n",
       "28                                                None  ...   \n",
       "29                                                None  ...   \n",
       "..                                                 ...  ...   \n",
       "251                                               None  ...   \n",
       "252                                               None  ...   \n",
       "253                                               None  ...   \n",
       "254                                               None  ...   \n",
       "255                                               None  ...   \n",
       "256                                               None  ...   \n",
       "257                                               None  ...   \n",
       "258                                               None  ...   \n",
       "259                                               None  ...   \n",
       "260                     http://jaxmice.jax.org/strain/  ...   \n",
       "261                     http://jaxmice.jax.org/strain/  ...   \n",
       "262  http://www.mmrrc.org/catalog/getSDS.jsp?mmrrc_id=  ...   \n",
       "263                     http://jaxmice.jax.org/strain/  ...   \n",
       "264                     http://jaxmice.jax.org/strain/  ...   \n",
       "265                     http://jaxmice.jax.org/strain/  ...   \n",
       "266                     http://jaxmice.jax.org/strain/  ...   \n",
       "267                     http://jaxmice.jax.org/strain/  ...   \n",
       "268                     http://jaxmice.jax.org/strain/  ...   \n",
       "269                     http://jaxmice.jax.org/strain/  ...   \n",
       "270                     http://jaxmice.jax.org/strain/  ...   \n",
       "271                     http://jaxmice.jax.org/strain/  ...   \n",
       "272                     http://jaxmice.jax.org/strain/  ...   \n",
       "273  http://www.mmrrc.org/catalog/getSDS.jsp?mmrrc_id=  ...   \n",
       "274                     http://jaxmice.jax.org/strain/  ...   \n",
       "275                     http://jaxmice.jax.org/strain/  ...   \n",
       "276  http://www.mmrrc.org/catalog/getSDS.jsp?mmrrc_id=  ...   \n",
       "277  http://www.mmrrc.org/catalog/getSDS.jsp?mmrrc_id=  ...   \n",
       "278  http://www.mmrrc.org/catalog/getSDS.jsp?mmrrc_id=  ...   \n",
       "279                     http://jaxmice.jax.org/strain/  ...   \n",
       "280                     http://jaxmice.jax.org/strain/  ...   \n",
       "\n",
       "                                           Driver Line  Abbreviation  \\\n",
       "0                                        Ctgf-2A-dgCre          Ctgf   \n",
       "1                                     Ndnf-IRES2-dgCre          Ndnf   \n",
       "2                                                  NaN           NaN   \n",
       "3    *Only used for isolation of cortical RNA for o...           NaN   \n",
       "4                                        Ctgf-2A-dgCre          Ctgf   \n",
       "5                                     Ndnf-IRES2-dgCre          Ndnf   \n",
       "6                                                  NaN           NaN   \n",
       "7    *Only used for isolation of cortical RNA for o...           NaN   \n",
       "8                                        Ctgf-2A-dgCre          Ctgf   \n",
       "9                                     Ndnf-IRES2-dgCre          Ndnf   \n",
       "10                                                 NaN           NaN   \n",
       "11   *Only used for isolation of cortical RNA for o...           NaN   \n",
       "12                                       Ctgf-2A-dgCre          Ctgf   \n",
       "13                                    Ndnf-IRES2-dgCre          Ndnf   \n",
       "14                                                 NaN           NaN   \n",
       "15   *Only used for isolation of cortical RNA for o...           NaN   \n",
       "16                                       Ctgf-2A-dgCre          Ctgf   \n",
       "17                                    Ndnf-IRES2-dgCre          Ndnf   \n",
       "18                                                 NaN           NaN   \n",
       "19   *Only used for isolation of cortical RNA for o...           NaN   \n",
       "20                                       Ctgf-2A-dgCre          Ctgf   \n",
       "21                                    Ndnf-IRES2-dgCre          Ndnf   \n",
       "22                                                 NaN           NaN   \n",
       "23   *Only used for isolation of cortical RNA for o...           NaN   \n",
       "24                                       Ctgf-2A-dgCre          Ctgf   \n",
       "25                                    Ndnf-IRES2-dgCre          Ndnf   \n",
       "26                                                 NaN           NaN   \n",
       "27   *Only used for isolation of cortical RNA for o...           NaN   \n",
       "28                                       Ctgf-2A-dgCre          Ctgf   \n",
       "29                                    Ndnf-IRES2-dgCre          Ndnf   \n",
       "..                                                 ...           ...   \n",
       "251  *Only used for isolation of cortical RNA for o...           NaN   \n",
       "252                                      Ctgf-2A-dgCre          Ctgf   \n",
       "253                                   Ndnf-IRES2-dgCre          Ndnf   \n",
       "254                                                NaN           NaN   \n",
       "255  *Only used for isolation of cortical RNA for o...           NaN   \n",
       "256                                      Ctgf-2A-dgCre          Ctgf   \n",
       "257                                   Ndnf-IRES2-dgCre          Ndnf   \n",
       "258                                                NaN           NaN   \n",
       "259  *Only used for isolation of cortical RNA for o...           NaN   \n",
       "260                                      Gad2-IRES-Cre          Gad2   \n",
       "261                                      Chat-IRES-Cre          Chat   \n",
       "262                                    Htr3a-Cre_NO152         Htr3a   \n",
       "263                                     Pvalb-IRES-Cre         Pvalb   \n",
       "264                                          Nr5a1-Cre         Nr5a1   \n",
       "265                                       Sst-IRES-Cre           Sst   \n",
       "266                                      Pvalb-2A-Flpo        PvalbF   \n",
       "267                                     Scnn1a-Tg3-Cre    Scnn1a-Tg3   \n",
       "268                                     Nkx2.1-CreERT2        Nkx2.1   \n",
       "269                                       Vip-IRES-Cre           Vip   \n",
       "270                                     Tac1-IRES2-Cre          Tac1   \n",
       "271                                     Calb2-IRES-Cre         Calb2   \n",
       "272                                   Trib2-2A-CreERT2        Trib2*   \n",
       "273                                       Cux2-CreERT2          Cux2   \n",
       "274                                     Scnn1a-Tg2-Cre    Scnn1a-Tg2   \n",
       "275                                       Pvalb-2A-Dre        PvalbD   \n",
       "276                                    Ntsr1-Cre_GN220         Ntsr1   \n",
       "277                                     Rbp4-Cre_KL100          Rbp4   \n",
       "278                                    Chrna2-Cre_OE25        Chrna2   \n",
       "279                                       Nos1-CreERT2          Nos1   \n",
       "280                                     Rorb-IRES2-Cre          Rorb   \n",
       "\n",
       "    Originating Lab (Donating Investigator)  \\\n",
       "0         Allen Institute for Brain Science   \n",
       "1         Allen Institute for Brain Science   \n",
       "2                                       NaN   \n",
       "3                                       NaN   \n",
       "4         Allen Institute for Brain Science   \n",
       "5         Allen Institute for Brain Science   \n",
       "6                                       NaN   \n",
       "7                                       NaN   \n",
       "8         Allen Institute for Brain Science   \n",
       "9         Allen Institute for Brain Science   \n",
       "10                                      NaN   \n",
       "11                                      NaN   \n",
       "12        Allen Institute for Brain Science   \n",
       "13        Allen Institute for Brain Science   \n",
       "14                                      NaN   \n",
       "15                                      NaN   \n",
       "16        Allen Institute for Brain Science   \n",
       "17        Allen Institute for Brain Science   \n",
       "18                                      NaN   \n",
       "19                                      NaN   \n",
       "20        Allen Institute for Brain Science   \n",
       "21        Allen Institute for Brain Science   \n",
       "22                                      NaN   \n",
       "23                                      NaN   \n",
       "24        Allen Institute for Brain Science   \n",
       "25        Allen Institute for Brain Science   \n",
       "26                                      NaN   \n",
       "27                                      NaN   \n",
       "28        Allen Institute for Brain Science   \n",
       "29        Allen Institute for Brain Science   \n",
       "..                                      ...   \n",
       "251                                     NaN   \n",
       "252       Allen Institute for Brain Science   \n",
       "253       Allen Institute for Brain Science   \n",
       "254                                     NaN   \n",
       "255                                     NaN   \n",
       "256       Allen Institute for Brain Science   \n",
       "257       Allen Institute for Brain Science   \n",
       "258                                     NaN   \n",
       "259                                     NaN   \n",
       "260                           Z. Josh Huang   \n",
       "261                             Brad Lowell   \n",
       "262    Nathaniel Heintz and Charles Gerfen    \n",
       "263                            Silvia Arber   \n",
       "264                             Brad Lowell   \n",
       "265                           Z. Josh Huang   \n",
       "266       Allen Institute for Brain Science   \n",
       "267       Allen Institute for Brain Science   \n",
       "268                           Z. Josh Huang   \n",
       "269                           Z. Josh Huang   \n",
       "270       Allen Institute for Brain Science   \n",
       "271                           Z. Josh Huang   \n",
       "272       Allen Institute for Brain Science   \n",
       "273                          Ulrich Mueller   \n",
       "274       Allen Institute for Brain Science   \n",
       "275       Allen Institute for Brain Science   \n",
       "276    Nathaniel Heintz and Charles Gerfen    \n",
       "277     Nathaniel Heintz and Charles Gerfen   \n",
       "278     Nathaniel Heintz and Charles Gerfen   \n",
       "279                           Z. Josh Huang   \n",
       "280       Allen Institute for Brain Science   \n",
       "\n",
       "                               Primary Reference Generation Method  \\\n",
       "0                                     This paper          Knock-in   \n",
       "1                                     This paper          Knock-in   \n",
       "2                                            NaN               NaN   \n",
       "3                                            NaN               NaN   \n",
       "4                                     This paper          Knock-in   \n",
       "5                                     This paper          Knock-in   \n",
       "6                                            NaN               NaN   \n",
       "7                                            NaN               NaN   \n",
       "8                                     This paper          Knock-in   \n",
       "9                                     This paper          Knock-in   \n",
       "10                                           NaN               NaN   \n",
       "11                                           NaN               NaN   \n",
       "12                                    This paper          Knock-in   \n",
       "13                                    This paper          Knock-in   \n",
       "14                                           NaN               NaN   \n",
       "15                                           NaN               NaN   \n",
       "16                                    This paper          Knock-in   \n",
       "17                                    This paper          Knock-in   \n",
       "18                                           NaN               NaN   \n",
       "19                                           NaN               NaN   \n",
       "20                                    This paper          Knock-in   \n",
       "21                                    This paper          Knock-in   \n",
       "22                                           NaN               NaN   \n",
       "23                                           NaN               NaN   \n",
       "24                                    This paper          Knock-in   \n",
       "25                                    This paper          Knock-in   \n",
       "26                                           NaN               NaN   \n",
       "27                                           NaN               NaN   \n",
       "28                                    This paper          Knock-in   \n",
       "29                                    This paper          Knock-in   \n",
       "..                                           ...               ...   \n",
       "251                                          NaN               NaN   \n",
       "252                                   This paper          Knock-in   \n",
       "253                                   This paper          Knock-in   \n",
       "254                                          NaN               NaN   \n",
       "255                                          NaN               NaN   \n",
       "256                                   This paper          Knock-in   \n",
       "257                                   This paper          Knock-in   \n",
       "258                                          NaN               NaN   \n",
       "259                                          NaN               NaN   \n",
       "260                Taniguchi et al., Neuron 2011          Knock-in   \n",
       "261               Rossi et al., Cell Metab. 2011          Knock-in   \n",
       "262                   Gerfen et al., Neuron 2013        Transgenic   \n",
       "263           Hippenmeyer et al., PLoS Biol 2005          Knock-in   \n",
       "264                  Dhillon et al., Neuron 2006        Transgenic   \n",
       "265                Taniguchi et al., Neuron 2011          Knock-in   \n",
       "266                  Madisen et al., Neuron 2015          Knock-in   \n",
       "267               Madisen et al., Nat. Neu. 2010        Transgenic   \n",
       "268               Taniguchi et al., Science 2013          Knock-in   \n",
       "269                Taniguchi et al., Neuron 2011          Knock-in   \n",
       "270  Harris et al., Front. Neural Circuits, 2014          Knock-in   \n",
       "271                Taniguchi et al., Neuron 2011          Knock-in   \n",
       "272  Harris et al., Front. Neural Circuits, 2014          Knock-in   \n",
       "273                  Franco et al., Science 2012          Knock-in   \n",
       "274               Madisen et al., Nat. Neu. 2010        Transgenic   \n",
       "275                  Madisen et al., Neuron 2015          Knock-in   \n",
       "276                   Gerfen et al., Neuron 2013        Transgenic   \n",
       "277                   Gerfen et al., Neuron 2013        Transgenic   \n",
       "278                   Gerfen et al., Neuron 2013        Transgenic   \n",
       "279                Taniguchi et al., Neuron 2011          Knock-in   \n",
       "280  Harris et al., Front. Neural Circuits, 2014          Knock-in   \n",
       "\n",
       "    Generation Method, more detail       Public Repository   \\\n",
       "0                              T2A                      NaN   \n",
       "1                            IRES2                      NaN   \n",
       "2                              NaN                      NaN   \n",
       "3                              NaN                      NaN   \n",
       "4                              T2A                      NaN   \n",
       "5                            IRES2                      NaN   \n",
       "6                              NaN                      NaN   \n",
       "7                              NaN                      NaN   \n",
       "8                              T2A                      NaN   \n",
       "9                            IRES2                      NaN   \n",
       "10                             NaN                      NaN   \n",
       "11                             NaN                      NaN   \n",
       "12                             T2A                      NaN   \n",
       "13                           IRES2                      NaN   \n",
       "14                             NaN                      NaN   \n",
       "15                             NaN                      NaN   \n",
       "16                             T2A                      NaN   \n",
       "17                           IRES2                      NaN   \n",
       "18                             NaN                      NaN   \n",
       "19                             NaN                      NaN   \n",
       "20                             T2A                      NaN   \n",
       "21                           IRES2                      NaN   \n",
       "22                             NaN                      NaN   \n",
       "23                             NaN                      NaN   \n",
       "24                             T2A                      NaN   \n",
       "25                           IRES2                      NaN   \n",
       "26                             NaN                      NaN   \n",
       "27                             NaN                      NaN   \n",
       "28                             T2A                      NaN   \n",
       "29                           IRES2                      NaN   \n",
       "..                             ...                      ...   \n",
       "251                            NaN                      NaN   \n",
       "252                            T2A                      NaN   \n",
       "253                          IRES2                      NaN   \n",
       "254                            NaN                      NaN   \n",
       "255                            NaN                      NaN   \n",
       "256                            T2A                      NaN   \n",
       "257                          IRES2                      NaN   \n",
       "258                            NaN                      NaN   \n",
       "259                            NaN                      NaN   \n",
       "260                           IRES   The Jackson Laboratory   \n",
       "261                           IRES   The Jackson Laboratory   \n",
       "262                            BAC                    MMRRC   \n",
       "263                           IRES   The Jackson Laboratory   \n",
       "264                            BAC   The Jackson Laboratory   \n",
       "265                           IRES   The Jackson Laboratory   \n",
       "266                            T2A  The Jackson Laboratory    \n",
       "267                            BAC  The Jackson Laboratory    \n",
       "268                         Direct   The Jackson Laboratory   \n",
       "269                           IRES   The Jackson Laboratory   \n",
       "270                          IRES2   The Jackson Laboratory   \n",
       "271                           IRES   The Jackson Laboratory   \n",
       "272                            F2A   The Jackson Laboratory   \n",
       "273                         Direct                    MMRRC   \n",
       "274                            BAC  The Jackson Laboratory    \n",
       "275                            T2A  The Jackson Laboratory    \n",
       "276                            BAC                    MMRRC   \n",
       "277                            BAC                    MMRRC   \n",
       "278                            BAC                    MMRRC   \n",
       "279                         Direct   The Jackson Laboratory   \n",
       "280                          IRES2   The Jackson Laboratory   \n",
       "\n",
       "    Public Repository Stock #                   Repository Strain Name  \\\n",
       "0                         NaN                                      NaN   \n",
       "1                         NaN                                      NaN   \n",
       "2                         NaN                                      NaN   \n",
       "3                         NaN                                      NaN   \n",
       "4                         NaN                                      NaN   \n",
       "5                         NaN                                      NaN   \n",
       "6                         NaN                                      NaN   \n",
       "7                         NaN                                      NaN   \n",
       "8                         NaN                                      NaN   \n",
       "9                         NaN                                      NaN   \n",
       "10                        NaN                                      NaN   \n",
       "11                        NaN                                      NaN   \n",
       "12                        NaN                                      NaN   \n",
       "13                        NaN                                      NaN   \n",
       "14                        NaN                                      NaN   \n",
       "15                        NaN                                      NaN   \n",
       "16                        NaN                                      NaN   \n",
       "17                        NaN                                      NaN   \n",
       "18                        NaN                                      NaN   \n",
       "19                        NaN                                      NaN   \n",
       "20                        NaN                                      NaN   \n",
       "21                        NaN                                      NaN   \n",
       "22                        NaN                                      NaN   \n",
       "23                        NaN                                      NaN   \n",
       "24                        NaN                                      NaN   \n",
       "25                        NaN                                      NaN   \n",
       "26                        NaN                                      NaN   \n",
       "27                        NaN                                      NaN   \n",
       "28                        NaN                                      NaN   \n",
       "29                        NaN                                      NaN   \n",
       "..                        ...                                      ...   \n",
       "251                       NaN                                      NaN   \n",
       "252                       NaN                                      NaN   \n",
       "253                       NaN                                      NaN   \n",
       "254                       NaN                                      NaN   \n",
       "255                       NaN                                      NaN   \n",
       "256                       NaN                                      NaN   \n",
       "257                       NaN                                      NaN   \n",
       "258                       NaN                                      NaN   \n",
       "259                       NaN                                      NaN   \n",
       "260                   10802.0                  STOCK Gad2tm2(cre)Zjh/J   \n",
       "261                    6410.0              B6;129S6-Chattm1(cre)Lowl/J   \n",
       "262                   36680.0       STOCK Tg(Htr3a-cre)NO152Gsat/Mmucd   \n",
       "263                    8069.0             B6;129P2-Pvalbtm1(cre)Arbr/J   \n",
       "264                    6364.0                 FVB-Tg(Nr5a1-cre)2Lowl/J   \n",
       "265                   13044.0                 STOCK Ssttm2.1(cre)Zjh/J   \n",
       "266                   22730.0              B6.Cg-Pvalbtm4.1(FLPo)Hze/J   \n",
       "267                    9613.0              B6;C3-Tg(Scnn1a-cre)3Aibs/J   \n",
       "268                   14552.0         STOCK Nkx2-1tm1.1(cre/ERT2)Zjh/J   \n",
       "269                   10908.0                   STOCK Viptm1(cre)Zjh/J   \n",
       "270                   21877.0              B6;129S-Tac1tm1.1(cre)Hze/J   \n",
       "271                   10774.0                B6(Cg)-Calb2tm1(cre)Zjh/J   \n",
       "272                   22865.0          B6.Cg-Trib2tm1.1(cre/ERT2)Hze/J   \n",
       "273                   32779.0    B6(Cg)-Cux2<tm3.1(cre/ERT2)Mull>/Mmmh   \n",
       "274                    9112.0              B6;C3-Tg(Scnn1a-cre)2Aibs/J   \n",
       "275                   21190.0              B6.Cg-Pvalbtm3.1(dreo)Hze/J   \n",
       "276                   30648.0  B6.FVB(Cg)-Tg(Ntsr1-cre)GN220Gsat/Mmucd   \n",
       "277                   31125.0        STOCK Tg(Rbp4-cre)KL100Gsat/Mmucd   \n",
       "278                   36502.0       STOCK Tg(Chrna2-cre)OE25Gsat/Mmucd   \n",
       "279                   14541.0         B6;129S-Nos1tm1.1(cre/ERT2)Zjh/J   \n",
       "280                   23526.0              B6;129S-Rorbtm1.1(cre)Hze/J   \n",
       "\n",
       "     Data available through the Allen Institute Transgenic Portal  \n",
       "0    http://connectivity.brain-map.org/transgenic/s...             \n",
       "1    http://connectivity.brain-map.org/transgenic/s...             \n",
       "2                                                  NaN             \n",
       "3                                                  NaN             \n",
       "4    http://connectivity.brain-map.org/transgenic/s...             \n",
       "5    http://connectivity.brain-map.org/transgenic/s...             \n",
       "6                                                  NaN             \n",
       "7                                                  NaN             \n",
       "8    http://connectivity.brain-map.org/transgenic/s...             \n",
       "9    http://connectivity.brain-map.org/transgenic/s...             \n",
       "10                                                 NaN             \n",
       "11                                                 NaN             \n",
       "12   http://connectivity.brain-map.org/transgenic/s...             \n",
       "13   http://connectivity.brain-map.org/transgenic/s...             \n",
       "14                                                 NaN             \n",
       "15                                                 NaN             \n",
       "16   http://connectivity.brain-map.org/transgenic/s...             \n",
       "17   http://connectivity.brain-map.org/transgenic/s...             \n",
       "18                                                 NaN             \n",
       "19                                                 NaN             \n",
       "20   http://connectivity.brain-map.org/transgenic/s...             \n",
       "21   http://connectivity.brain-map.org/transgenic/s...             \n",
       "22                                                 NaN             \n",
       "23                                                 NaN             \n",
       "24   http://connectivity.brain-map.org/transgenic/s...             \n",
       "25   http://connectivity.brain-map.org/transgenic/s...             \n",
       "26                                                 NaN             \n",
       "27                                                 NaN             \n",
       "28   http://connectivity.brain-map.org/transgenic/s...             \n",
       "29   http://connectivity.brain-map.org/transgenic/s...             \n",
       "..                                                 ...             \n",
       "251                                                NaN             \n",
       "252  http://connectivity.brain-map.org/transgenic/s...             \n",
       "253  http://connectivity.brain-map.org/transgenic/s...             \n",
       "254                                                NaN             \n",
       "255                                                NaN             \n",
       "256  http://connectivity.brain-map.org/transgenic/s...             \n",
       "257  http://connectivity.brain-map.org/transgenic/s...             \n",
       "258                                                NaN             \n",
       "259                                                NaN             \n",
       "260  http://connectivity.brain-map.org/transgenic/s...             \n",
       "261  http://connectivity.brain-map.org/transgenic/s...             \n",
       "262  http://connectivity.brain-map.org/transgenic/s...             \n",
       "263  http://connectivity.brain-map.org/transgenic/s...             \n",
       "264  http://connectivity.brain-map.org/transgenic/s...             \n",
       "265  http://connectivity.brain-map.org/transgenic/s...             \n",
       "266  http://connectivity.brain-map.org/transgenic/s...             \n",
       "267  http://connectivity.brain-map.org/transgenic/s...             \n",
       "268  http://connectivity.brain-map.org/transgenic/s...             \n",
       "269  http://connectivity.brain-map.org/transgenic/s...             \n",
       "270  http://connectivity.brain-map.org/transgenic/s...             \n",
       "271  http://connectivity.brain-map.org/transgenic/s...             \n",
       "272  http://connectivity.brain-map.org/transgenic/s...             \n",
       "273  http://connectivity.brain-map.org/transgenic/s...             \n",
       "274  http://connectivity.brain-map.org/transgenic/s...             \n",
       "275  http://connectivity.brain-map.org/transgenic/s...             \n",
       "276  http://connectivity.brain-map.org/transgenic/s...             \n",
       "277  http://connectivity.brain-map.org/transgenic/s...             \n",
       "278  http://connectivity.brain-map.org/transgenic/s...             \n",
       "279  http://connectivity.brain-map.org/transgenic/s...             \n",
       "280  http://connectivity.brain-map.org/transgenic/s...             \n",
       "\n",
       "[281 rows x 22 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cre_df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cre_df = cre_df[['Abbreviation','name', 'id', 'stock_number',\n",
    "                 'transgenic_line_source_name',\n",
    "                 'transgenic_line_type_name',\n",
    "                 'url_prefix', 'url_suffix','description']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cre_df2 = pd.merge(df_cre_mtd, cre_ref,  how='inner', \n",
    "                   left_on='Public Repository Stock #',\n",
    "                   right_on = 'stock_number')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyontutils.namespaces import makePrefixes, ilxtr, definition\n",
    "from pyontutils.closed_namespaces import rdf, rdfs, owl\n",
    "from pyontutils.config import devconfig\n",
    "\n",
    "class Tasic2015(Genes, phns.Species, \n",
    "                phns.Regions, phns.Layers):\n",
    "    branch = devconfig.neurons_branch\n",
    "    \n",
    "    #TODO: Ambiguous, most likely post-clustering layers\n",
    "    L6a = Phenotype(ilxtr.TasicL6a, ilxtr.hasSomaLocatedIn, \n",
    "                    label = \"Tasic Layer VI - A\", override = True)\n",
    "    L6b = Phenotype(ilxtr.TasicL6b, ilxtr.hasSomaLocatedIn, \n",
    "                    label = \"Tasic Layer VI - B\", override = True)\n",
    "    \n",
    "    #Aggregate Layer Phenotypes\n",
    "    with phns.Layers:\n",
    "        upper = LogicalPhenotype(OR, L1, L2, L3)\n",
    "        lower = LogicalPhenotype(OR, L4, L5, L6)\n",
    "        All = LogicalPhenotype(OR, upper, lower)\n",
    "    \n",
    "    #Tasic Hiearchical Cluster Position (1-indexed, breadth first)\n",
    "    cmp  = pred.hasComputedMolecularPhenotype\n",
    "\n",
    "class TasicBagger:\n",
    "    #from Allen Cell Types\n",
    "    prefixes = {**{'JAX': 'http://jaxmice.jax.org/strain/',\n",
    "                   'MMRRC': 'http://www.mmrrc.org/catalog/getSDS.jsp?mmrrc_id=',\n",
    "                   'AllenTL': 'http://api.brain-map.org/api/v2/data/TransgenicLine/'},\n",
    "                **makePrefixes('definition', 'ilxtr', 'owl')}\n",
    "\n",
    "    def __init__(self, data = None, **metadata):\n",
    "        \"\"\"\n",
    "        Initializes a Tasic Neuron Bagger Object\n",
    "        \n",
    "        data: Pandas DataFrame\n",
    "        **metadata: any relevant metadata DataFrames\n",
    "        \"\"\"\n",
    "        self.data = data\n",
    "        self.ns = {k:rdflib.Namespace(v) for k, v in self.prefixes.items()}\n",
    "        if metadata:\n",
    "            for key, dataframe in metadata.items():\n",
    "                setattr(self, key, dataframe)\n",
    "            \n",
    "    @staticmethod\n",
    "    def layer_parse(key):\n",
    "        \"\"\"method that parses the layer labels for Tasic 2015 cells.\n",
    "        returns the layer phenotype.\n",
    "        \n",
    "        key: str, layer label\n",
    "        \"\"\"\n",
    "        if key == 'L2/3':\n",
    "            layer_phn = L23     \n",
    "        else:\n",
    "            layer_phn = Tasic2015[key]\n",
    "        \n",
    "        return layer_phn\n",
    "    \n",
    "    @staticmethod\n",
    "    def gene_parse(gene_set, mode = 'present'):\n",
    "        \"\"\"method that parses a set of gene markers and returns\n",
    "        a list of phenotypes.\n",
    "        \n",
    "        gene_set: set, set of gene markers in dtype str\n",
    "        mode: str, \"present\" or \"absent\" markers.\n",
    "        \"\"\"\n",
    "        gene_phns = []\n",
    "        undef = set()\n",
    "        f = Phenotype if mode == 'present' else NegPhenotype\n",
    "        for gene in gene_set:\n",
    "            if gene in Genes.__dict__:\n",
    "                gene_phns.append(f(Genes[gene]))\n",
    "            else:\n",
    "                undef.add(gene)\n",
    "        if len(undef) > 0: #FIXME: some genes cannot be mapped\n",
    "            mappings, to_add, errors = ncbigenemapping(undef, \n",
    "                                        return_errors = True)\n",
    "            for gene_name, iri in mappings.items():\n",
    "                gene_phns.append(f(iri, ilxtr.hasExpressionPhenotype,\n",
    "                                   label = gene_name, override = True))\n",
    "        if len(gene_phns) != len(gene_set):\n",
    "            print(errors)\n",
    "        return gene_phns\n",
    "    \n",
    "    @staticmethod\n",
    "    def cluster_parse(pos):\n",
    "        \"\"\"Tasic Computed Gene-based hiearchical cluster\n",
    "        pos: int\n",
    "        \"\"\"\n",
    "        cluster_phn = Phenotype(\"ilxtr:cluster\" + str(pos), cmp, \n",
    "                                label = str(pos))\n",
    "        return cluster_phn\n",
    "    \n",
    "    \n",
    "    def transgenic_parse(self, row):\n",
    "        phenotypes = []\n",
    "        pred = 'ilxtr:hasDriverExpressionPhenotype'\n",
    "        cre = self.cre_metadata[self.cre_metadata['Abbreviation'] == row['cre']]\n",
    "        cre_phn = Phenotype('PR:000013502', pred.hasExpressionPhenotype)\n",
    "        ##WHERE R THE IRIs ??\n",
    "        #for tl in cell_line['donor']['transgenic_lines']:\n",
    "            #prefix = self.cre_metadata['Driver Line']\n",
    "            #suffix = self.cre_metadata['Public Repository Stock #'] if tl['stock_number'] else str(tl['id'])\n",
    "            #line_names = []\n",
    "            #if prefix and suffix and prefix in ['AIBS', 'MMRRC', 'JAX']:\n",
    "                #if prefix == 'AIBS':\n",
    "                   # prefix = 'AllenTL'\n",
    "               # iri = self.ns[prefix][suffix]\n",
    "               # phenotypes.append(Phenotype(iri, pred))\n",
    "        return cre_phn\n",
    "    \n",
    "    def build_transgenic_lines(self):\n",
    "        triples = []\n",
    "        for ind, tl in self.cre.iterrows():\n",
    "            _id = tl['stock_number'] if tl['stock_number'] else tl['id']\n",
    "            prefix = tl['transgenic_line_source_name']\n",
    "            line_type = tl['transgenic_line_type_name']\n",
    "            if prefix not in ['JAX', 'MMRRC', 'AIBS']:\n",
    "                print('WARNING:', 'unknown prefix')\n",
    "                continue\n",
    "            elif prefix == 'AIBS':\n",
    "                prefix = 'AllenTL'\n",
    "\n",
    "            _class = self.ns[prefix][str(_id)]\n",
    "            triples.append((_class, rdf.type, owl.Class))\n",
    "            triples.append((_class, rdfs.label, rdflib.Literal(tl['name'])))\n",
    "            triples.append((_class, definition, rdflib.Literal(tl['description'])))\n",
    "            triples.append((_class, rdfs.subClassOf, ilxtr.transgenicLine))\n",
    "            triples.append((_class, ilxtr.hasTransgenicType, ilxtr[line_type + 'Line']))\n",
    "\n",
    "        # TODO aspects.ttl?\n",
    "        transgenic_lines = simpleOnt(filename='tasic-transgenic-lines',\n",
    "                                     path='ttl/generated/',\n",
    "                                     prefixes=self.prefixes,\n",
    "                                     triples=triples,\n",
    "                                     comment='Tasic transgenic lines for cell types',\n",
    "                                     branch=self.branch)\n",
    "\n",
    "        transgenic_lines._graph.write()\n",
    "        \n",
    "        #return transgenic_lines   \n",
    "    \n",
    "    @property\n",
    "    def bags(self):\n",
    "        with Tasic2015:\n",
    "            #Every neuron sampled in this paper is from V1\n",
    "            with Neuron(phns.Species.Mouse, phns.Regions.V1) as context:\n",
    "                for row in self.data.itertuples():\n",
    "                    label = str(row.short_name) #change this\n",
    "                    #cluster_phn = Phenotype(str(row.cluster), cmp)\n",
    "                    #layer_phn = TasicBagger.layer_parse(row.layer_dissectoin)\n",
    "\n",
    "                    present_phns = TasicBagger.gene_parse(row.markers_present, mode = 'present')\n",
    "                    absent_phns  = TasicBagger.gene_parse(row.markers_absent, mode = 'absent')\n",
    "                    markers_phns = present_phns + absent_phns\n",
    "\n",
    "                    #phenotypes = [layer_phn, cre_phn, cluster_phn] + markers_phns\n",
    "                    phenotypes = markers_phns\n",
    "                    yield label, phenotypes\n",
    "          \n",
    "    \n",
    "class TasicNeuron(NeuronEBM):\n",
    "    owlClass = ilxtr.NeuronTasic2015\n",
    "    shortname = 'Tasic2015'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = {\"cre\":cre_df}\n",
    "\n",
    "def main():\n",
    "    ttl_test_path = '/var/host/media/removable/SD Card/Neuron/Tasic/ttl_export'\n",
    "    config = Config(\"tasic-2015\", ttl_export_dir=Path(ttl_test_path))\n",
    "    tb = TasicBagger(data = df_types, **metadata)\n",
    "    ind = 0\n",
    "    for label, bag in tb.bags:\n",
    "        if ind == 2:\n",
    "            break\n",
    "        TasicNeuron(*bag, label = label, override=True)\n",
    "        ind += 1\n",
    "    config.write()\n",
    "    config.write_python()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "<module '__main__'> is a built-in class",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-104-f63e976de425>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mconfig\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite_python\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/var/host/media/removable/SD Card/Neuron/pyontutils/neurondm/neurondm/core.py\u001b[0m in \u001b[0;36mwrite_python\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    662\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mwrite_python\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    663\u001b[0m         \u001b[0;31m# FIXME hack, will write other configs if call after graphbase has switched\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 664\u001b[0;31m         \u001b[0mgraphBase\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite_python\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    665\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    666\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mload_existing\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/var/host/media/removable/SD Card/Neuron/pyontutils/neurondm/neurondm/core.py\u001b[0m in \u001b[0;36mwrite_python\u001b[0;34m()\u001b[0m\n\u001b[1;32m   1090\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mstaticmethod\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1091\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mwrite_python\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1092\u001b[0;31m         \u001b[0mpython\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgraphBase\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1093\u001b[0m         \u001b[0;31m# if you try to read from a source file that already exists\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1094\u001b[0m         \u001b[0;31m# while also writing to that file linecache will be smart and\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/var/host/media/removable/SD Card/Neuron/pyontutils/neurondm/neurondm/core.py\u001b[0m in \u001b[0;36mpython\u001b[0;34m(cls)\u001b[0m\n\u001b[1;32m   1133\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mclassmethod\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1134\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mpython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcls\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1135\u001b[0;31m         \u001b[0mout\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcls\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpython_header\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1136\u001b[0m         \u001b[0;31m#out += '\\n\\n'.join('\\n'.join(('# ' + n.label, '# ' + n._origLabel, str(n))) for n in neurons)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1137\u001b[0m         \u001b[0mout\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;34m'\\n\\n'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mn\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcls\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mneurons\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m# FIXME this does not reset correctly when a new Controller is created, it probably should...\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/var/host/media/removable/SD Card/Neuron/pyontutils/neurondm/neurondm/core.py\u001b[0m in \u001b[0;36mpython_header\u001b[0;34m(cls)\u001b[0m\n\u001b[1;32m   1105\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1106\u001b[0m         \u001b[0mall_types\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mn\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcls\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mneurons\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1107\u001b[0;31m         _subs = [inspect.getsource(c) for c in subclasses(Neuron)\n\u001b[0m\u001b[1;32m   1108\u001b[0m                  if c in all_types and Path(inspect.getfile(c)).exists()]\n\u001b[1;32m   1109\u001b[0m         \u001b[0msubs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'\\n'\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m'\\n\\n'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_subs\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m'\\n\\n'\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0m_subs\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/var/host/media/removable/SD Card/Neuron/pyontutils/neurondm/neurondm/core.py\u001b[0m in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m   1106\u001b[0m         \u001b[0mall_types\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mn\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcls\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mneurons\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1107\u001b[0m         _subs = [inspect.getsource(c) for c in subclasses(Neuron)\n\u001b[0;32m-> 1108\u001b[0;31m                  if c in all_types and Path(inspect.getfile(c)).exists()]\n\u001b[0m\u001b[1;32m   1109\u001b[0m         \u001b[0msubs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'\\n'\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m'\\n\\n'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_subs\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m'\\n\\n'\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0m_subs\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1110\u001b[0m         \u001b[0;31m#log.debug(str(all_types))\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/neuron/lib/python3.7/inspect.py\u001b[0m in \u001b[0;36mgetfile\u001b[0;34m(object)\u001b[0m\n\u001b[1;32m    651\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobject\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'__file__'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    652\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__file__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 653\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'{!r} is a built-in class'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobject\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    654\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mismethod\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobject\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    655\u001b[0m         \u001b[0mobject\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__func__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: <module '__main__'> is a built-in class"
     ]
    }
   ],
   "source": [
    "config.write_python()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('22004', 'Tpm2', '10090'), ('57266', 'Cxcl14', '10090'), ('209378', 'Itih5', '10090'), ('12862', 'Cox6a2', '10090')]\n",
      "[('20660', 'Sorl1', '10090'), ('12351', 'Car4', '10090')]\n",
      "[('12351', 'Car4', '10090'), ('21334', 'Tac2', '10090')]\n",
      "[('18167', 'Npy2r', '10090'), ('242259', 'Slc44a5', '10090')]\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "<module '__main__'> is a built-in class",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-99-a051907d3f74>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      9\u001b[0m     \u001b[0mind\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0mconfig\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 11\u001b[0;31m \u001b[0mconfig\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite_python\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/var/host/media/removable/SD Card/Neuron/pyontutils/neurondm/neurondm/core.py\u001b[0m in \u001b[0;36mwrite_python\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    662\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mwrite_python\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    663\u001b[0m         \u001b[0;31m# FIXME hack, will write other configs if call after graphbase has switched\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 664\u001b[0;31m         \u001b[0mgraphBase\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite_python\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    665\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    666\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mload_existing\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/var/host/media/removable/SD Card/Neuron/pyontutils/neurondm/neurondm/core.py\u001b[0m in \u001b[0;36mwrite_python\u001b[0;34m()\u001b[0m\n\u001b[1;32m   1090\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mstaticmethod\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1091\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mwrite_python\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1092\u001b[0;31m         \u001b[0mpython\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgraphBase\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1093\u001b[0m         \u001b[0;31m# if you try to read from a source file that already exists\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1094\u001b[0m         \u001b[0;31m# while also writing to that file linecache will be smart and\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/var/host/media/removable/SD Card/Neuron/pyontutils/neurondm/neurondm/core.py\u001b[0m in \u001b[0;36mpython\u001b[0;34m(cls)\u001b[0m\n\u001b[1;32m   1133\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mclassmethod\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1134\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mpython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcls\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1135\u001b[0;31m         \u001b[0mout\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcls\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpython_header\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1136\u001b[0m         \u001b[0;31m#out += '\\n\\n'.join('\\n'.join(('# ' + n.label, '# ' + n._origLabel, str(n))) for n in neurons)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1137\u001b[0m         \u001b[0mout\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;34m'\\n\\n'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mn\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcls\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mneurons\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m# FIXME this does not reset correctly when a new Controller is created, it probably should...\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/var/host/media/removable/SD Card/Neuron/pyontutils/neurondm/neurondm/core.py\u001b[0m in \u001b[0;36mpython_header\u001b[0;34m(cls)\u001b[0m\n\u001b[1;32m   1105\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1106\u001b[0m         \u001b[0mall_types\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mn\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcls\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mneurons\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1107\u001b[0;31m         _subs = [inspect.getsource(c) for c in subclasses(Neuron)\n\u001b[0m\u001b[1;32m   1108\u001b[0m                  if c in all_types and Path(inspect.getfile(c)).exists()]\n\u001b[1;32m   1109\u001b[0m         \u001b[0msubs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'\\n'\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m'\\n\\n'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_subs\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m'\\n\\n'\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0m_subs\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/var/host/media/removable/SD Card/Neuron/pyontutils/neurondm/neurondm/core.py\u001b[0m in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m   1106\u001b[0m         \u001b[0mall_types\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mn\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcls\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mneurons\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1107\u001b[0m         _subs = [inspect.getsource(c) for c in subclasses(Neuron)\n\u001b[0;32m-> 1108\u001b[0;31m                  if c in all_types and Path(inspect.getfile(c)).exists()]\n\u001b[0m\u001b[1;32m   1109\u001b[0m         \u001b[0msubs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'\\n'\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m'\\n\\n'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_subs\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m'\\n\\n'\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0m_subs\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1110\u001b[0m         \u001b[0;31m#log.debug(str(all_types))\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/neuron/lib/python3.7/inspect.py\u001b[0m in \u001b[0;36mgetfile\u001b[0;34m(object)\u001b[0m\n\u001b[1;32m    651\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobject\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'__file__'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    652\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__file__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 653\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'{!r} is a built-in class'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobject\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    654\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mismethod\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobject\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    655\u001b[0m         \u001b[0mobject\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__func__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: <module '__main__'> is a built-in class"
     ]
    }
   ],
   "source": [
    "ttl_test_path = '/var/host/media/removable/SD Card/Neuron/Tasic/ttl_export'\n",
    "config = Config(\"tasic-2015\", ttl_export_dir=Path(ttl_test_path))\n",
    "tb = TasicBagger(data = df_types, **metadata)\n",
    "ind = 0\n",
    "for label, bag in tb.bags:\n",
    "    if ind == 1:\n",
    "        break\n",
    "    TasicNeuron(*bag, label = label, override=True)\n",
    "    ind += 1\n",
    "config.write()\n",
    "config.write_python()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('22004', 'Tpm2', '10090'), ('57266', 'Cxcl14', '10090'), ('209378', 'Itih5', '10090'), ('12862', 'Cox6a2', '10090')]\n",
      "[('20660', 'Sorl1', '10090'), ('12351', 'Car4', '10090')]\n",
      "[('12351', 'Car4', '10090'), ('21334', 'Tac2', '10090')]\n",
      "[('18167', 'Npy2r', '10090'), ('242259', 'Slc44a5', '10090')]\n",
      "[('73750', 'Whrn', '10090'), ('18573', 'Pde1a', '10090'), ('16323', 'Inhba', '10090'), ('13602', 'Sparcl1', '10090'), ('109593', 'Lmo3', '10090'), ('225998', 'Rorb', '10090')]\n",
      "[('320265', 'Tafa1', '10090'), ('17311', 'Kitl', '10090'), ('19011', 'Endou', '10090')]\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "<module '__main__'> is a built-in class",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-97-05a752170ec0>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mpyontutils\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclosed_namespaces\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mrdfs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m \u001b[0mmain\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-96-0f1b5270cff7>\u001b[0m in \u001b[0;36mmain\u001b[0;34m()\u001b[0m\n\u001b[1;32m     12\u001b[0m         \u001b[0mind\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m     \u001b[0mconfig\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 14\u001b[0;31m     \u001b[0mconfig\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite_python\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/var/host/media/removable/SD Card/Neuron/pyontutils/neurondm/neurondm/core.py\u001b[0m in \u001b[0;36mwrite_python\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    662\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mwrite_python\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    663\u001b[0m         \u001b[0;31m# FIXME hack, will write other configs if call after graphbase has switched\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 664\u001b[0;31m         \u001b[0mgraphBase\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite_python\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    665\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    666\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mload_existing\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/var/host/media/removable/SD Card/Neuron/pyontutils/neurondm/neurondm/core.py\u001b[0m in \u001b[0;36mwrite_python\u001b[0;34m()\u001b[0m\n\u001b[1;32m   1090\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mstaticmethod\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1091\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mwrite_python\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1092\u001b[0;31m         \u001b[0mpython\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgraphBase\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1093\u001b[0m         \u001b[0;31m# if you try to read from a source file that already exists\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1094\u001b[0m         \u001b[0;31m# while also writing to that file linecache will be smart and\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/var/host/media/removable/SD Card/Neuron/pyontutils/neurondm/neurondm/core.py\u001b[0m in \u001b[0;36mpython\u001b[0;34m(cls)\u001b[0m\n\u001b[1;32m   1133\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mclassmethod\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1134\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mpython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcls\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1135\u001b[0;31m         \u001b[0mout\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcls\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpython_header\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1136\u001b[0m         \u001b[0;31m#out += '\\n\\n'.join('\\n'.join(('# ' + n.label, '# ' + n._origLabel, str(n))) for n in neurons)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1137\u001b[0m         \u001b[0mout\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;34m'\\n\\n'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mn\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcls\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mneurons\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m# FIXME this does not reset correctly when a new Controller is created, it probably should...\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/var/host/media/removable/SD Card/Neuron/pyontutils/neurondm/neurondm/core.py\u001b[0m in \u001b[0;36mpython_header\u001b[0;34m(cls)\u001b[0m\n\u001b[1;32m   1105\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1106\u001b[0m         \u001b[0mall_types\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mn\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcls\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mneurons\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1107\u001b[0;31m         _subs = [inspect.getsource(c) for c in subclasses(Neuron)\n\u001b[0m\u001b[1;32m   1108\u001b[0m                  if c in all_types and Path(inspect.getfile(c)).exists()]\n\u001b[1;32m   1109\u001b[0m         \u001b[0msubs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'\\n'\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m'\\n\\n'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_subs\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m'\\n\\n'\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0m_subs\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/var/host/media/removable/SD Card/Neuron/pyontutils/neurondm/neurondm/core.py\u001b[0m in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m   1106\u001b[0m         \u001b[0mall_types\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mn\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcls\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mneurons\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1107\u001b[0m         _subs = [inspect.getsource(c) for c in subclasses(Neuron)\n\u001b[0;32m-> 1108\u001b[0;31m                  if c in all_types and Path(inspect.getfile(c)).exists()]\n\u001b[0m\u001b[1;32m   1109\u001b[0m         \u001b[0msubs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'\\n'\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m'\\n\\n'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_subs\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m'\\n\\n'\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0m_subs\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1110\u001b[0m         \u001b[0;31m#log.debug(str(all_types))\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/neuron/lib/python3.7/inspect.py\u001b[0m in \u001b[0;36mgetfile\u001b[0;34m(object)\u001b[0m\n\u001b[1;32m    651\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobject\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'__file__'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    652\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__file__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 653\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'{!r} is a built-in class'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobject\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    654\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mismethod\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobject\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    655\u001b[0m         \u001b[0mobject\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__func__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: <module '__main__'> is a built-in class"
     ]
    }
   ],
   "source": [
    "from neurondm.sheets import Sheet\n",
    "from neurondm import OntId, OntTerm, Config, NeuronEBM, Neuron\n",
    "from pyontutils.utils import byCol, relative_path\n",
    "from pyontutils.namespaces import ilxtr\n",
    "from pyontutils.closed_namespaces import rdfs\n",
    "\n",
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = {\"cre\":cre_df}\n",
    "tb = TasicBagger(data = df_types, **metadata)\n",
    "tb.build_transgenic_lines()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test Code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Phenotype('NCBITaxon:10090',\n",
      "          'ilxtr:hasInstanceInSpecies',\n",
      "          label='Mus musculus')\n"
     ]
    }
   ],
   "source": [
    "with Tasic2015:\n",
    "    print(Mouse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyontutils.core import simpleOnt\n",
    "simpleOnt()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Phenotype(\"ilxtr:cluster6\", \"ilxtr:hasComputedMolecularPhenotype\", label = \"Tasic\", check = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyontutils.utils import byCol, relative_path\n",
    "from pyontutils.namespaces import ilxtr\n",
    "from pyontutils.closed_namespaces import rdfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with Neuron(phns.Species.Mouse, phns.Regions.V1) as context:\n",
    "    print(context)\n",
    "    n11 = Neuron(phns.Layers.L1)\n",
    "    print(n11)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
