{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TODO (project-level): \n",
    "\n",
    "- Generation of Cell names\n",
    "- names for higher subclasses\n",
    "- penotypic & Genotypic -- Check threshold\n",
    "  \n",
    "----------------------------------------------------\n",
    "\n",
    "- Start building from individual neurons\n",
    "- Naming scheme: Tasic 15, cluster [cluster_number]\n",
    "- Function that changes names flexibly\n",
    "- Templates & References\n",
    "    - Bolser Lewis \n",
    "    - OntTerm triplesimple\n",
    "    - nistd/core.py\n",
    "    - curation.py = triplesExport\n",
    "    \n",
    "#### Code-level:\n",
    "- Start using templates from NeuronLangExample\n",
    "- Recapitulating the matrix with definitive phenotypes\n",
    "- populate namespace with definitive phenotypesi\n",
    "\n",
    "Definitive phenotype:\n",
    "    - Cre line [NCBI Taxon] [Allen Brain Institute]\n",
    "    - V1 \n",
    "    - Layer of dissection [UBERON]\n",
    "\n",
    "Inferred gene markers phenotype\n",
    "    - Cluster ID [ILX]\n",
    "    - Genes present [NCBIGene]\n",
    "    - Genes Absent [NCBIGene]\n",
    "\n",
    "#Tasic hiearchy\n",
    "### For the Future:\n",
    "- Continuous Phenotypes (gene counts, etc)\n",
    "- Phenotype Inheritance in subclasses?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "http://casestudies.brain-map.org/celltax\n",
    "\n",
    "https://www.nature.com/articles/nn.4216"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "#standard libraries\n",
    "from pathlib import Path\n",
    "from importlib import reload\n",
    "\n",
    "#extended libraries\n",
    "import requests\n",
    "import rdflib\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy as sp\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "#tree stuff\n",
    "import anytree\n",
    "import anytree.util\n",
    "from anytree.importer import DictImporter\n",
    "from anytree.exporter import DotExporter\n",
    "\n",
    "#custom imports\n",
    "from neurondm import *\n",
    "from neurondm.models.huang2017 import Genes\n",
    "from pyontutils.namespaces import ilxtr as pred\n",
    "from neurondm import phenotype_namespaces as phns\n",
    "from pyontutils.closed_namespaces import rdf, rdfs, owl\n",
    "#from nifstd.nifstd_tools.utils import ncbigenemapping\n",
    "from utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{}"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pred.__dict__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cellguard(addns=False):\n",
    "    # unfortunately ipy.hooks['pre_run_code_hook'].add(__cellguard)\n",
    "    # causes this to be called too frequently :/\n",
    "    setLocalNames()\n",
    "    setLocalContext()\n",
    "    if addns:\n",
    "        setLocalNames(phns.BBP)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_clf = pd.read_csv('Data/cell_classification.csv')\n",
    "df_cls_mtd = pd.read_csv('Data/cluster_metadata.csv')\n",
    "df_cell_mtd = pd.read_csv('Data/cell_metadata.csv')\n",
    "\n",
    "df_cre_mtd  = pd.read_excel('Data/tasic_crelines.xlsx')\n",
    "\n",
    "dendro = pd.read_csv(\"Data/web/big_tree_final.csv\", dtype = {'position':str})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Information about the cluster membership of each cell,\n",
    "including whether the cell is a \"core\" (unambiguously assigned to a single cluster) \n",
    "or \"transition\" (shares membership between two clusters) cell, \n",
    "as well as its membership score (from 0-10) for each cluster (labeled f01 to f49). \n",
    "\"\"\"\n",
    "df_clf.head()\n",
    "#TODO: Only including unambiguous cells with f-values = 10 for one cluster only."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    " Information about each data-driven cluster, including its label, \n",
    " the corresponding label in Tasic et al. (Nat. Neuro, 2106), the primary cell class \n",
    " membership, and marker genes (including genes with widespread expression in the cluster, \n",
    " sparse expression in the cluster, and no expression in the cluster). \n",
    "\"\"\"\n",
    "df_cls_mtd.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Information about each cell profiled, including its nomenclature, \n",
    "Cre line of origin, dissection, date of collection and sequencing, \n",
    "and read mapping statistics\n",
    "\"\"\"\n",
    "df_cell_mtd.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_clf['coretype'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cell_mtd['layer_dissectoin'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cell_mtd['cre'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_clf.rename(index=str, columns={\"Unnamed: 0\":\"cell_index\"}, inplace = True)\n",
    "df_clf = df_clf[['cell_index','coretype', 'primary', 'secondary']]\n",
    "\n",
    "#Reorder and drop columns that are not interesting\n",
    "df_cell_mtd = df_cell_mtd[['short_name','cre','major_class','sub_class',\n",
    "                           'major_dissection', 'layer_dissectoin']]\n",
    "df_cls_mtd = df_cls_mtd[['cluster_id','cluster_order','vignette_label',\n",
    "                         'group','markers_present','markers_sparse',\n",
    "                         'genes_absent','Tasic_et_al_2016_label']]\n",
    "\n",
    "df_cls_mtd['cluster_id'] = df_cls_mtd['cluster_id'].astype(str)\n",
    "\n",
    "#merge\n",
    "df_types = df_cell_mtd.merge(df_clf, left_on='short_name', right_on='cell_index')\n",
    "df_types = df_types[['short_name', 'coretype', 'primary', 'secondary',\n",
    "                     'cre','major_dissection', 'layer_dissectoin']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cre line conversion\n",
    "df_cre_mtd.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "df_types.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# metada query test\n",
    "df_cre_mtd[df_cre_mtd['Abbreviation'] == df_types['cre'][1000]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "main dataframe: df_types\n",
    "\n",
    "metadata: cre_mtd & cls_mtd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dendrogram Parsing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cd34</td>\n",
       "      <td>000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ndnf Car4</td>\n",
       "      <td>0000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ndnf Cxcl14</td>\n",
       "      <td>0000011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Vip Gpc3</td>\n",
       "      <td>0000100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Vip Chat</td>\n",
       "      <td>0000101</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       cluster position\n",
       "0         Cd34   000000\n",
       "1    Ndnf Car4  0000010\n",
       "2  Ndnf Cxcl14  0000011\n",
       "3     Vip Gpc3  0000100\n",
       "4     Vip Chat  0000101"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dendro.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "tree_dict = {'label':'root'} #base dictionary\n",
    "\n",
    "#parses binary node positions into a dictionary with tree structure\n",
    "for label, bin_str in zip(dendro['cluster'], dendro['position']):\n",
    "    parse_binary(tree_dict, bin_str, label)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnyNode(label='root', name='1', pos=1)\n",
      "├── AnyNode(name='2', pos=2)\n",
      "│   ├── AnyNode(name='4', pos=4)\n",
      "│   │   ├── AnyNode(name='8', pos=8)\n",
      "│   │   │   ├── AnyNode(name='16', pos=16)\n",
      "│   │   │   │   ├── AnyNode(name='28', pos=28)\n",
      "│   │   │   │   │   ├── AnyNode(label='Cd34', name='42 Cd34', pos=42)\n",
      "│   │   │   │   │   └── AnyNode(name='43', pos=43)\n",
      "│   │   │   │   │       ├── AnyNode(label='Ndnf Car4', name='60 Ndnf Car4', pos=60)\n",
      "│   │   │   │   │       └── AnyNode(label='Ndnf Cxcl14', name='61 Ndnf Cxcl14', pos=61)\n",
      "│   │   │   │   └── AnyNode(name='29', pos=29)\n",
      "│   │   │   │       ├── AnyNode(name='44', pos=44)\n",
      "│   │   │   │       │   ├── AnyNode(label='Vip Gpc3', name='62 Vip Gpc3', pos=62)\n",
      "│   │   │   │       │   └── AnyNode(label='Vip Chat', name='63 Vip Chat', pos=63)\n",
      "│   │   │   │       └── AnyNode(name='45', pos=45)\n",
      "│   │   │   │           ├── AnyNode(name='64', pos=64)\n",
      "│   │   │   │           │   ├── AnyNode(label='Vip Cxcl14_Car4', name='80 Vip Cxcl14_Car4', pos=80)\n",
      "│   │   │   │           │   └── AnyNode(label='Vip Mybpc1', name='81 Vip Mybpc1', pos=81)\n",
      "│   │   │   │           └── AnyNode(label='Vip Sncg', name='65 Vip Sncg', pos=65)\n",
      "│   │   │   └── AnyNode(label='Sncg', name='17 Sncg', pos=17)\n",
      "│   │   └── AnyNode(name='9', pos=9)\n",
      "│   │       ├── AnyNode(name='18', pos=18)\n",
      "│   │       │   ├── AnyNode(name='30', pos=30)\n",
      "│   │       │   │   ├── AnyNode(label='Sst Chodl', name='46 Sst Chodl', pos=46)\n",
      "│   │       │   │   └── AnyNode(name='47', pos=47)\n",
      "│   │       │   │       ├── AnyNode(name='66', pos=66)\n",
      "│   │       │   │       │   ├── AnyNode(name='82', pos=82)\n",
      "│   │       │   │       │   │   ├── AnyNode(label='Sst Cdk6', name='88 Sst Cdk6', pos=88)\n",
      "│   │       │   │       │   │   └── AnyNode(name='89', pos=89)\n",
      "│   │       │   │       │   │       ├── AnyNode(label='Sst Cbln4', name='94 Sst Cbln4', pos=94)\n",
      "│   │       │   │       │   │       └── AnyNode(label='Sst Myh8', name='95 Sst Myh8', pos=95)\n",
      "│   │       │   │       │   └── AnyNode(label='Sst Tacstd2', name='83 Sst Tacstd2', pos=83)\n",
      "│   │       │   │       └── AnyNode(label='Sst Th', name='67 Sst Th', pos=67)\n",
      "│   │       │   └── AnyNode(name='31', pos=31)\n",
      "│   │       │       ├── AnyNode(label='Pvalb Cpne5', name='48 Pvalb Cpne5', pos=48)\n",
      "│   │       │       └── AnyNode(name='49', pos=49)\n",
      "│   │       │           ├── AnyNode(label='Pvalb Tpbg', name='68 Pvalb Tpbg', pos=68)\n",
      "│   │       │           └── AnyNode(name='69', pos=69)\n",
      "│   │       │               ├── AnyNode(name='84', pos=84)\n",
      "│   │       │               │   ├── AnyNode(label='Pvalb Gpx3', name='90 Pvalb Gpx3', pos=90)\n",
      "│   │       │               │   └── AnyNode(name='91', pos=91)\n",
      "│   │       │               │       ├── AnyNode(label='Pvalb Rspo2', name='96 Pvalb Rspo2', pos=96)\n",
      "│   │       │               │       └── AnyNode(label='Pvalb Wt1', name='97 Pvalb Wt1', pos=97)\n",
      "│   │       │               └── AnyNode(name='85', pos=85)\n",
      "│   │       │                   ├── AnyNode(label='Pvalb Obox3', name='92 Pvalb Obox3', pos=92)\n",
      "│   │       │                   └── AnyNode(label='Pvalb Tacr3', name='93 Pvalb Tacr3', pos=93)\n",
      "│   │       └── AnyNode(label='Igtp', name='19 Igtp', pos=19)\n",
      "│   └── AnyNode(name='5', pos=5)\n",
      "│       ├── AnyNode(name='10', pos=10)\n",
      "│       │   ├── AnyNode(name='20', pos=20)\n",
      "│       │   │   ├── AnyNode(label='L2 Ngb', name='32 L2 Ngb', pos=32)\n",
      "│       │   │   └── AnyNode(label='L2/3 Otof', name='33 L2/3 Otof', pos=33)\n",
      "│       │   └── AnyNode(name='21', pos=21)\n",
      "│       │       ├── AnyNode(name='34', pos=34)\n",
      "│       │       │   ├── AnyNode(name='50', pos=50)\n",
      "│       │       │   │   ├── AnyNode(label='L4 Sparcl1', name='70 L4 Sparcl1', pos=70)\n",
      "│       │       │   │   └── AnyNode(label='L4 Scnn1a', name='71 L4 Scnn1a', pos=71)\n",
      "│       │       │   └── AnyNode(label='L4 Arf5', name='51 L4 Arf5', pos=51)\n",
      "│       │       └── AnyNode(label='L4 Hsd11b1', name='35 L4 Hsd11b1', pos=35)\n",
      "│       └── AnyNode(name='11', pos=11)\n",
      "│           ├── AnyNode(name='22', pos=22)\n",
      "│           │   ├── AnyNode(name='36', pos=36)\n",
      "│           │   │   ├── AnyNode(name='52', pos=52)\n",
      "│           │   │   │   ├── AnyNode(label='L5a Fam5c', name='72 L5a Fam5c', pos=72)\n",
      "│           │   │   │   └── AnyNode(name='73', pos=73)\n",
      "│           │   │   │       ├── AnyNode(label='L5a Syt17', name='86 L5a Syt17', pos=86)\n",
      "│           │   │   │       └── AnyNode(label='L5a Batf3', name='87 L5a Batf3', pos=87)\n",
      "│           │   │   └── AnyNode(label='L5 Ucma', name='53 L5 Ucma', pos=53)\n",
      "│           │   └── AnyNode(name='37', pos=37)\n",
      "│           │       ├── AnyNode(name='54', pos=54)\n",
      "│           │       │   ├── AnyNode(label='L5b Samd3', name='74 L5b Samd3', pos=74)\n",
      "│           │       │   └── AnyNode(label='L5b Cdh13', name='75 L5b Cdh13', pos=75)\n",
      "│           │       └── AnyNode(label='L5 Chrna6', name='55 L5 Chrna6', pos=55)\n",
      "│           └── AnyNode(name='23', pos=23)\n",
      "│               ├── AnyNode(name='38', pos=38)\n",
      "│               │   ├── AnyNode(name='56', pos=56)\n",
      "│               │   │   ├── AnyNode(label='L6a Mgp', name='76 L6a Mgp', pos=76)\n",
      "│               │   │   └── AnyNode(label='L6a Plcxd3', name='77 L6a Plcxd3', pos=77)\n",
      "│               │   └── AnyNode(name='57', pos=57)\n",
      "│               │       ├── AnyNode(label='L6a Car12', name='78 L6a Car12', pos=78)\n",
      "│               │       └── AnyNode(label='L6a Syt17', name='79 L6a Syt17', pos=79)\n",
      "│               └── AnyNode(name='39', pos=39)\n",
      "│                   ├── AnyNode(label='L6b Trh', name='58 L6b Trh', pos=58)\n",
      "│                   └── AnyNode(label='L6b Mup5', name='59 L6b Mup5', pos=59)\n",
      "└── AnyNode(name='3', pos=3)\n",
      "    ├── AnyNode(name='6', pos=6)\n",
      "    │   ├── AnyNode(label='Astro Gja1', name='12 Astro Gja1', pos=12)\n",
      "    │   └── AnyNode(name='13', pos=13)\n",
      "    │       ├── AnyNode(label='OPC Pdgfra', name='24 OPC Pdgfra', pos=24)\n",
      "    │       └── AnyNode(name='25', pos=25)\n",
      "    │           ├── AnyNode(label='OPC/Oligo', name='40 OPC/Oligo', pos=40)\n",
      "    │           └── AnyNode(label='Oligo Opalin', name='41 Oligo Opalin', pos=41)\n",
      "    └── AnyNode(name='7', pos=7)\n",
      "        ├── AnyNode(label='Micro Ctss', name='14 Micro Ctss', pos=14)\n",
      "        └── AnyNode(name='15', pos=15)\n",
      "            ├── AnyNode(label='Endo Tbc1d4', name='26 Endo Tbc1d4', pos=26)\n",
      "            └── AnyNode(label='Endo Myl9', name='27 Endo Myl9', pos=27)\n"
     ]
    }
   ],
   "source": [
    "importer = DictImporter()\n",
    "tree = importer.import_(tree_dict)\n",
    "\n",
    "for ind, node in enumerate(anytree.LevelOrderIter(tree)):\n",
    "    node.pos = ind + 1 #node index starting at one\n",
    "    if node.is_leaf:\n",
    "        node.name = str(ind + 1) + \" \" + node.label\n",
    "    else:\n",
    "        node.name = str(ind + 1)\n",
    "        \n",
    "print(anytree.RenderTree(tree))\n",
    "DotExporter(tree).to_picture(\"dendro.png\")\n",
    "\n",
    "#node.name for rendering\n",
    "#node.pos & node.label used for parsing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "for leaf in tree.leaves: #TODO: phenotypes & deal with mismatching\n",
    "    #edge cases for the last two endothelial cells\n",
    "    if leaf.label  == 'Endo Tbc1d4':\n",
    "        leaf.cluster_id = 'f48'\n",
    "    elif leaf.label == 'Endo Myl9':\n",
    "        leaf.cluster_id = 'f49'\n",
    "    else:\n",
    "        leaf.cluster_id = list(df_cls_mtd[df_cls_mtd['vignette_label'] == leaf.label]['cluster_id'])[0]\n",
    "    \n",
    "    #FIXME: label name matching somehow not working\n",
    "    \"\"\"\n",
    "    genes_present = list(df_cls_mtd[df_cls_mtd['cluster_id'] == leaf.cluster_id]['markers_present'].str.split(\",\"))\n",
    "    genes_absent = list(df_cls_mtd[df_cls_mtd['cluster_id'] == leaf.cluster_id]['genes_absent'].str.split(\",\"))\n",
    "    if genes_present == np.nan:\n",
    "        leaf.genes_present = set()\n",
    "    if genes_absent == np.nan:\n",
    "        leaf.genes_absent = set()\n",
    "    else:\n",
    "        leaf.genes_present = set(genes_present[0])\n",
    "        #leaf.genes_absent = set(genes_absent)\n",
    "    \"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gene and cluster columns converted from dendrogram structure\n",
    "df_types['cluster'] = df_types.apply(cluster_converter, tree = tree, axis = 1)\n",
    "df_types['markers_present'] = df_types.apply(gene_merge, df = df_cls_mtd, index = 'markers_present', axis = 1)\n",
    "df_types['markers_absent'] = df_types.apply(gene_merge, df = df_cls_mtd, index = 'genes_absent', axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>short_name</th>\n",
       "      <th>coretype</th>\n",
       "      <th>primary</th>\n",
       "      <th>secondary</th>\n",
       "      <th>cre</th>\n",
       "      <th>major_dissection</th>\n",
       "      <th>layer_dissectoin</th>\n",
       "      <th>cluster</th>\n",
       "      <th>markers_present</th>\n",
       "      <th>markers_absent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A272_V</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f24</td>\n",
       "      <td>f28</td>\n",
       "      <td>Calb2</td>\n",
       "      <td>V1</td>\n",
       "      <td>All</td>\n",
       "      <td>10</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ Gm13629, Kcnab1,  Fam19a1,  Galntl6,  Endou,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>A279_V</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f07</td>\n",
       "      <td>f02</td>\n",
       "      <td>Calb2</td>\n",
       "      <td>V1</td>\n",
       "      <td>All</td>\n",
       "      <td>16</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Mybpc1,  Npy2r,  Slc44a5, Car4}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>A1612_V</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f32</td>\n",
       "      <td>f39</td>\n",
       "      <td>Calb2</td>\n",
       "      <td>V1</td>\n",
       "      <td>All</td>\n",
       "      <td>11</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ Pter, Ntrk3,  Slc30a3, C1ql3}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>A1613_V</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f49</td>\n",
       "      <td>f44</td>\n",
       "      <td>Calb2</td>\n",
       "      <td>V1</td>\n",
       "      <td>All</td>\n",
       "      <td>3</td>\n",
       "      <td>{}</td>\n",
       "      <td>{AI467606}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>A1619_VU</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f02</td>\n",
       "      <td>f07</td>\n",
       "      <td>Calb2</td>\n",
       "      <td>V1</td>\n",
       "      <td>upper</td>\n",
       "      <td>16</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Mybpc1,  Npy2r,  Slc44a5, Car4}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>A1621_VU</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f07</td>\n",
       "      <td>f05</td>\n",
       "      <td>Calb2</td>\n",
       "      <td>V1</td>\n",
       "      <td>upper</td>\n",
       "      <td>16</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Car4}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>A1622_VU</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f01</td>\n",
       "      <td>f02</td>\n",
       "      <td>Calb2</td>\n",
       "      <td>V1</td>\n",
       "      <td>upper</td>\n",
       "      <td>64</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Mybpc1,  Car4,  Npy2r,  Slc44a5,  Sorl1, Npy2r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>A1623_VU</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f02</td>\n",
       "      <td>f01</td>\n",
       "      <td>Calb2</td>\n",
       "      <td>V1</td>\n",
       "      <td>upper</td>\n",
       "      <td>64</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Mybpc1,  Npy2r,  Car4,  Slc44a5,  Sorl1, Npy2r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>A1626_VU</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f06</td>\n",
       "      <td>f07</td>\n",
       "      <td>Calb2</td>\n",
       "      <td>V1</td>\n",
       "      <td>upper</td>\n",
       "      <td>43</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Car4, Runx1t1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>A1629_VU</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f07</td>\n",
       "      <td>f05</td>\n",
       "      <td>Calb2</td>\n",
       "      <td>V1</td>\n",
       "      <td>upper</td>\n",
       "      <td>16</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Car4}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>A1636_VL</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f28</td>\n",
       "      <td>f24</td>\n",
       "      <td>Calb2</td>\n",
       "      <td>V1</td>\n",
       "      <td>lower</td>\n",
       "      <td>10</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ Gm13629, Kcnab1,  Fam19a1,  Galntl6,  Endou,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>A1645_VL</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f02</td>\n",
       "      <td>f01</td>\n",
       "      <td>Calb2</td>\n",
       "      <td>V1</td>\n",
       "      <td>lower</td>\n",
       "      <td>64</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Mybpc1,  Npy2r,  Car4,  Slc44a5,  Sorl1, Npy2r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>A1646_VL</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f49</td>\n",
       "      <td>f44</td>\n",
       "      <td>Calb2</td>\n",
       "      <td>V1</td>\n",
       "      <td>lower</td>\n",
       "      <td>3</td>\n",
       "      <td>{}</td>\n",
       "      <td>{AI467606}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>B600_V4</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f26</td>\n",
       "      <td>f30</td>\n",
       "      <td>Rorb</td>\n",
       "      <td>V1</td>\n",
       "      <td>L4</td>\n",
       "      <td>5</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Rspo1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>B605_V4</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f25</td>\n",
       "      <td>f24</td>\n",
       "      <td>Rorb</td>\n",
       "      <td>V1</td>\n",
       "      <td>L4</td>\n",
       "      <td>50</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Scnn1a,  Fam19a1,  Kitl,  Endou}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>B822_V4</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f25</td>\n",
       "      <td>f24</td>\n",
       "      <td>Rorb</td>\n",
       "      <td>V1</td>\n",
       "      <td>L4</td>\n",
       "      <td>50</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Scnn1a,  Fam19a1,  Kitl,  Endou}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>B825_V4</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f25</td>\n",
       "      <td>f24</td>\n",
       "      <td>Rorb</td>\n",
       "      <td>V1</td>\n",
       "      <td>L4</td>\n",
       "      <td>50</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Scnn1a,  Fam19a1,  Kitl,  Endou}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>B827_V4</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f25</td>\n",
       "      <td>f24</td>\n",
       "      <td>Rorb</td>\n",
       "      <td>V1</td>\n",
       "      <td>L4</td>\n",
       "      <td>50</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Scnn1a,  Fam19a1,  Kitl,  Endou}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>B81_V4</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f27</td>\n",
       "      <td>f25</td>\n",
       "      <td>Rorb</td>\n",
       "      <td>V1</td>\n",
       "      <td>L4</td>\n",
       "      <td>34</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>B83_V4</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f25</td>\n",
       "      <td>f24</td>\n",
       "      <td>Rorb</td>\n",
       "      <td>V1</td>\n",
       "      <td>L4</td>\n",
       "      <td>50</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Scnn1a,  Fam19a1,  Kitl,  Endou}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>B85_V4</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f27</td>\n",
       "      <td>f25</td>\n",
       "      <td>Rorb</td>\n",
       "      <td>V1</td>\n",
       "      <td>L4</td>\n",
       "      <td>34</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>B161_V4</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f25</td>\n",
       "      <td>f24</td>\n",
       "      <td>Rorb</td>\n",
       "      <td>V1</td>\n",
       "      <td>L4</td>\n",
       "      <td>50</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Scnn1a,  Fam19a1,  Kitl,  Endou}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>B162_V4</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f25</td>\n",
       "      <td>f27</td>\n",
       "      <td>Rorb</td>\n",
       "      <td>V1</td>\n",
       "      <td>L4</td>\n",
       "      <td>34</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>B165_V4</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f25</td>\n",
       "      <td>f24</td>\n",
       "      <td>Rorb</td>\n",
       "      <td>V1</td>\n",
       "      <td>L4</td>\n",
       "      <td>50</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Scnn1a,  Fam19a1,  Kitl,  Endou}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>B390_V5</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f32</td>\n",
       "      <td>f31</td>\n",
       "      <td>Rorb</td>\n",
       "      <td>V1</td>\n",
       "      <td>L5</td>\n",
       "      <td>73</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Fam5c, Ntrk3}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>C188_V4</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f25</td>\n",
       "      <td>f24</td>\n",
       "      <td>Cux2</td>\n",
       "      <td>V1</td>\n",
       "      <td>L4</td>\n",
       "      <td>50</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Scnn1a,  Fam19a1,  Kitl,  Endou}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>C191_V4</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f24</td>\n",
       "      <td>f25</td>\n",
       "      <td>Cux2</td>\n",
       "      <td>V1</td>\n",
       "      <td>L4</td>\n",
       "      <td>50</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Scnn1a,  Fam19a1,  Kitl,  Endou}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>C489_V4</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f24</td>\n",
       "      <td>f25</td>\n",
       "      <td>Cux2</td>\n",
       "      <td>V1</td>\n",
       "      <td>L4</td>\n",
       "      <td>50</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Scnn1a,  Fam19a1,  Kitl,  Endou}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>C490_V4</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f25</td>\n",
       "      <td>f24</td>\n",
       "      <td>Cux2</td>\n",
       "      <td>V1</td>\n",
       "      <td>L4</td>\n",
       "      <td>50</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Scnn1a,  Fam19a1,  Kitl,  Endou}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>C493_V4</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f25</td>\n",
       "      <td>f24</td>\n",
       "      <td>Cux2</td>\n",
       "      <td>V1</td>\n",
       "      <td>L4</td>\n",
       "      <td>50</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Scnn1a,  Fam19a1,  Kitl,  Endou}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1463</th>\n",
       "      <td>U1471_V</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f31</td>\n",
       "      <td>f32</td>\n",
       "      <td>Nos1</td>\n",
       "      <td>V1</td>\n",
       "      <td>All</td>\n",
       "      <td>73</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Fam5c, Ntrk3}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1464</th>\n",
       "      <td>U1472_V</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f29</td>\n",
       "      <td>f28</td>\n",
       "      <td>Nos1</td>\n",
       "      <td>V1</td>\n",
       "      <td>All</td>\n",
       "      <td>20</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ Rnd3,  Gm13629,  Rcan2, Kcnab1,  Zmat4,  Gal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1467</th>\n",
       "      <td>U1475_V</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f31</td>\n",
       "      <td>f32</td>\n",
       "      <td>Nos1</td>\n",
       "      <td>V1</td>\n",
       "      <td>All</td>\n",
       "      <td>73</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Fam5c, Ntrk3}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1472</th>\n",
       "      <td>U1481_V</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f31</td>\n",
       "      <td>f32</td>\n",
       "      <td>Nos1</td>\n",
       "      <td>V1</td>\n",
       "      <td>All</td>\n",
       "      <td>73</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Fam5c, Ntrk3}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1493</th>\n",
       "      <td>U1507_V</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f24</td>\n",
       "      <td>f28</td>\n",
       "      <td>Nos1</td>\n",
       "      <td>V1</td>\n",
       "      <td>All</td>\n",
       "      <td>10</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ Gm13629, Kcnab1,  Fam19a1,  Galntl6,  Endou,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1498</th>\n",
       "      <td>V739_VU</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f02</td>\n",
       "      <td>f01</td>\n",
       "      <td>Vip</td>\n",
       "      <td>V1</td>\n",
       "      <td>upper</td>\n",
       "      <td>64</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Mybpc1,  Npy2r,  Car4,  Slc44a5,  Sorl1, Npy2r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1512</th>\n",
       "      <td>V754_VU</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f01</td>\n",
       "      <td>f02</td>\n",
       "      <td>Vip</td>\n",
       "      <td>V1</td>\n",
       "      <td>upper</td>\n",
       "      <td>64</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Mybpc1,  Car4,  Npy2r,  Slc44a5,  Sorl1, Npy2r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1514</th>\n",
       "      <td>V756_VU</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f01</td>\n",
       "      <td>f02</td>\n",
       "      <td>Vip</td>\n",
       "      <td>V1</td>\n",
       "      <td>upper</td>\n",
       "      <td>64</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Mybpc1,  Car4,  Npy2r,  Slc44a5,  Sorl1, Npy2r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1515</th>\n",
       "      <td>V757_VU</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f05</td>\n",
       "      <td>f02</td>\n",
       "      <td>Vip</td>\n",
       "      <td>V1</td>\n",
       "      <td>upper</td>\n",
       "      <td>29</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Mybpc1,  Npy2r,  Slc44a5}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1533</th>\n",
       "      <td>V776_VL</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f05</td>\n",
       "      <td>f04</td>\n",
       "      <td>Vip</td>\n",
       "      <td>V1</td>\n",
       "      <td>lower</td>\n",
       "      <td>44</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Crh,  Pnoc}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1537</th>\n",
       "      <td>V781_VL</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f05</td>\n",
       "      <td>f02</td>\n",
       "      <td>Vip</td>\n",
       "      <td>V1</td>\n",
       "      <td>lower</td>\n",
       "      <td>29</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Mybpc1,  Npy2r,  Slc44a5}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1542</th>\n",
       "      <td>V911_VL</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f05</td>\n",
       "      <td>f02</td>\n",
       "      <td>Vip</td>\n",
       "      <td>V1</td>\n",
       "      <td>lower</td>\n",
       "      <td>29</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Mybpc1,  Npy2r,  Slc44a5}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1567</th>\n",
       "      <td>W1798_VU</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f20</td>\n",
       "      <td>f08</td>\n",
       "      <td>Nkx2-1</td>\n",
       "      <td>V1</td>\n",
       "      <td>upper</td>\n",
       "      <td>9</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Cxcl14,  Arhgef3}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1568</th>\n",
       "      <td>W1799_VU</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f18</td>\n",
       "      <td>f17</td>\n",
       "      <td>Nkx2-1</td>\n",
       "      <td>V1</td>\n",
       "      <td>upper</td>\n",
       "      <td>49</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ Drd1a,  Tacr3, Tcap,  Cnr1, Col25a1,  Il1rap...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1570</th>\n",
       "      <td>W1801_VU</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f18</td>\n",
       "      <td>f20</td>\n",
       "      <td>Nkx2-1</td>\n",
       "      <td>V1</td>\n",
       "      <td>upper</td>\n",
       "      <td>49</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ Drd1a,  Pcsk2, Tcap,  Cnr1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1579</th>\n",
       "      <td>W1810_VU</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f22</td>\n",
       "      <td>f20</td>\n",
       "      <td>Nkx2-1</td>\n",
       "      <td>V1</td>\n",
       "      <td>upper</td>\n",
       "      <td>69</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Tac1,  Kirrel3,  Etv1,  Snca,  Zcchc12}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1595</th>\n",
       "      <td>Y1911_VU</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f22</td>\n",
       "      <td>f20</td>\n",
       "      <td>Pvalb</td>\n",
       "      <td>V1</td>\n",
       "      <td>upper</td>\n",
       "      <td>69</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Tac1,  Kirrel3,  Etv1,  Snca,  Zcchc12}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1598</th>\n",
       "      <td>Y1914_VU</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f18</td>\n",
       "      <td>f20</td>\n",
       "      <td>Pvalb</td>\n",
       "      <td>V1</td>\n",
       "      <td>upper</td>\n",
       "      <td>49</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ Drd1a,  Pcsk2, Tcap,  Cnr1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1603</th>\n",
       "      <td>Y1919_VU</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f18</td>\n",
       "      <td>f20</td>\n",
       "      <td>Pvalb</td>\n",
       "      <td>V1</td>\n",
       "      <td>upper</td>\n",
       "      <td>49</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ Drd1a,  Pcsk2, Tcap,  Cnr1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1608</th>\n",
       "      <td>Y1924_VL</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f17</td>\n",
       "      <td>f21</td>\n",
       "      <td>Pvalb</td>\n",
       "      <td>V1</td>\n",
       "      <td>lower</td>\n",
       "      <td>84</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ Tacr3, Col25a1,  Kcnd2,  Gap43,  Il1rapl2, E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1613</th>\n",
       "      <td>Y1929_VL</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f17</td>\n",
       "      <td>f22</td>\n",
       "      <td>Pvalb</td>\n",
       "      <td>V1</td>\n",
       "      <td>lower</td>\n",
       "      <td>84</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Tac1,  Kirrel3,  Tacr3, Col25a1,  Il1rapl2,  ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1614</th>\n",
       "      <td>Y1930_VL</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f17</td>\n",
       "      <td>f21</td>\n",
       "      <td>Pvalb</td>\n",
       "      <td>V1</td>\n",
       "      <td>lower</td>\n",
       "      <td>84</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ Tacr3, Col25a1,  Kcnd2,  Gap43,  Il1rapl2, E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1621</th>\n",
       "      <td>Y1937_VL</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f17</td>\n",
       "      <td>f20</td>\n",
       "      <td>Pvalb</td>\n",
       "      <td>V1</td>\n",
       "      <td>lower</td>\n",
       "      <td>69</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ Il1rapl2,  Tacr3, Col25a1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1626</th>\n",
       "      <td>Y1942_VL</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f22</td>\n",
       "      <td>f21</td>\n",
       "      <td>Pvalb</td>\n",
       "      <td>V1</td>\n",
       "      <td>lower</td>\n",
       "      <td>91</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Tac1,  Kirrel3,  Kcnd2,  Gap43,  Etv1,  Snca,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1627</th>\n",
       "      <td>Y1943_VL</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f21</td>\n",
       "      <td>f17</td>\n",
       "      <td>Pvalb</td>\n",
       "      <td>V1</td>\n",
       "      <td>lower</td>\n",
       "      <td>84</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ Kcnd2,  Gap43,  Tacr3, Col25a1,  Il1rapl2, E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1634</th>\n",
       "      <td>Y2054_VU</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f20</td>\n",
       "      <td>f22</td>\n",
       "      <td>Pvalb</td>\n",
       "      <td>V1</td>\n",
       "      <td>upper</td>\n",
       "      <td>69</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Tac1,  Kirrel3,  Etv1,  Snca,  Zcchc12}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1641</th>\n",
       "      <td>Y2061_VL</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f21</td>\n",
       "      <td>f17</td>\n",
       "      <td>Pvalb</td>\n",
       "      <td>V1</td>\n",
       "      <td>lower</td>\n",
       "      <td>84</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ Kcnd2,  Gap43,  Tacr3, Col25a1,  Il1rapl2, E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1643</th>\n",
       "      <td>Y2063_VL</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f20</td>\n",
       "      <td>f17</td>\n",
       "      <td>Pvalb</td>\n",
       "      <td>V1</td>\n",
       "      <td>lower</td>\n",
       "      <td>69</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ Il1rapl2,  Tacr3, Col25a1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1648</th>\n",
       "      <td>Y1429_V</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f22</td>\n",
       "      <td>f20</td>\n",
       "      <td>Pvalb</td>\n",
       "      <td>V1</td>\n",
       "      <td>All</td>\n",
       "      <td>69</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Tac1,  Kirrel3,  Etv1,  Snca,  Zcchc12}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1651</th>\n",
       "      <td>Y1432_V</td>\n",
       "      <td>Transition</td>\n",
       "      <td>f22</td>\n",
       "      <td>f17</td>\n",
       "      <td>Pvalb</td>\n",
       "      <td>V1</td>\n",
       "      <td>All</td>\n",
       "      <td>84</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Tac1,  Kirrel3,  Tacr3, Col25a1,  Etv1,  Snca...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>255 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     short_name    coretype primary secondary     cre major_dissection  \\\n",
       "8        A272_V  Transition     f24       f28   Calb2               V1   \n",
       "15       A279_V  Transition     f07       f02   Calb2               V1   \n",
       "26      A1612_V  Transition     f32       f39   Calb2               V1   \n",
       "27      A1613_V  Transition     f49       f44   Calb2               V1   \n",
       "33     A1619_VU  Transition     f02       f07   Calb2               V1   \n",
       "35     A1621_VU  Transition     f07       f05   Calb2               V1   \n",
       "36     A1622_VU  Transition     f01       f02   Calb2               V1   \n",
       "37     A1623_VU  Transition     f02       f01   Calb2               V1   \n",
       "40     A1626_VU  Transition     f06       f07   Calb2               V1   \n",
       "43     A1629_VU  Transition     f07       f05   Calb2               V1   \n",
       "50     A1636_VL  Transition     f28       f24   Calb2               V1   \n",
       "59     A1645_VL  Transition     f02       f01   Calb2               V1   \n",
       "60     A1646_VL  Transition     f49       f44   Calb2               V1   \n",
       "73      B600_V4  Transition     f26       f30    Rorb               V1   \n",
       "78      B605_V4  Transition     f25       f24    Rorb               V1   \n",
       "80      B822_V4  Transition     f25       f24    Rorb               V1   \n",
       "83      B825_V4  Transition     f25       f24    Rorb               V1   \n",
       "85      B827_V4  Transition     f25       f24    Rorb               V1   \n",
       "88       B81_V4  Transition     f27       f25    Rorb               V1   \n",
       "90       B83_V4  Transition     f25       f24    Rorb               V1   \n",
       "92       B85_V4  Transition     f27       f25    Rorb               V1   \n",
       "96      B161_V4  Transition     f25       f24    Rorb               V1   \n",
       "97      B162_V4  Transition     f25       f27    Rorb               V1   \n",
       "99      B165_V4  Transition     f25       f24    Rorb               V1   \n",
       "113     B390_V5  Transition     f32       f31    Rorb               V1   \n",
       "118     C188_V4  Transition     f25       f24    Cux2               V1   \n",
       "121     C191_V4  Transition     f24       f25    Cux2               V1   \n",
       "152     C489_V4  Transition     f24       f25    Cux2               V1   \n",
       "153     C490_V4  Transition     f25       f24    Cux2               V1   \n",
       "156     C493_V4  Transition     f25       f24    Cux2               V1   \n",
       "...         ...         ...     ...       ...     ...              ...   \n",
       "1463    U1471_V  Transition     f31       f32    Nos1               V1   \n",
       "1464    U1472_V  Transition     f29       f28    Nos1               V1   \n",
       "1467    U1475_V  Transition     f31       f32    Nos1               V1   \n",
       "1472    U1481_V  Transition     f31       f32    Nos1               V1   \n",
       "1493    U1507_V  Transition     f24       f28    Nos1               V1   \n",
       "1498    V739_VU  Transition     f02       f01     Vip               V1   \n",
       "1512    V754_VU  Transition     f01       f02     Vip               V1   \n",
       "1514    V756_VU  Transition     f01       f02     Vip               V1   \n",
       "1515    V757_VU  Transition     f05       f02     Vip               V1   \n",
       "1533    V776_VL  Transition     f05       f04     Vip               V1   \n",
       "1537    V781_VL  Transition     f05       f02     Vip               V1   \n",
       "1542    V911_VL  Transition     f05       f02     Vip               V1   \n",
       "1567   W1798_VU  Transition     f20       f08  Nkx2-1               V1   \n",
       "1568   W1799_VU  Transition     f18       f17  Nkx2-1               V1   \n",
       "1570   W1801_VU  Transition     f18       f20  Nkx2-1               V1   \n",
       "1579   W1810_VU  Transition     f22       f20  Nkx2-1               V1   \n",
       "1595   Y1911_VU  Transition     f22       f20   Pvalb               V1   \n",
       "1598   Y1914_VU  Transition     f18       f20   Pvalb               V1   \n",
       "1603   Y1919_VU  Transition     f18       f20   Pvalb               V1   \n",
       "1608   Y1924_VL  Transition     f17       f21   Pvalb               V1   \n",
       "1613   Y1929_VL  Transition     f17       f22   Pvalb               V1   \n",
       "1614   Y1930_VL  Transition     f17       f21   Pvalb               V1   \n",
       "1621   Y1937_VL  Transition     f17       f20   Pvalb               V1   \n",
       "1626   Y1942_VL  Transition     f22       f21   Pvalb               V1   \n",
       "1627   Y1943_VL  Transition     f21       f17   Pvalb               V1   \n",
       "1634   Y2054_VU  Transition     f20       f22   Pvalb               V1   \n",
       "1641   Y2061_VL  Transition     f21       f17   Pvalb               V1   \n",
       "1643   Y2063_VL  Transition     f20       f17   Pvalb               V1   \n",
       "1648    Y1429_V  Transition     f22       f20   Pvalb               V1   \n",
       "1651    Y1432_V  Transition     f22       f17   Pvalb               V1   \n",
       "\n",
       "     layer_dissectoin  cluster markers_present  \\\n",
       "8                 All       10              {}   \n",
       "15                All       16              {}   \n",
       "26                All       11              {}   \n",
       "27                All        3              {}   \n",
       "33              upper       16              {}   \n",
       "35              upper       16              {}   \n",
       "36              upper       64              {}   \n",
       "37              upper       64              {}   \n",
       "40              upper       43              {}   \n",
       "43              upper       16              {}   \n",
       "50              lower       10              {}   \n",
       "59              lower       64              {}   \n",
       "60              lower        3              {}   \n",
       "73                 L4        5              {}   \n",
       "78                 L4       50              {}   \n",
       "80                 L4       50              {}   \n",
       "83                 L4       50              {}   \n",
       "85                 L4       50              {}   \n",
       "88                 L4       34              {}   \n",
       "90                 L4       50              {}   \n",
       "92                 L4       34              {}   \n",
       "96                 L4       50              {}   \n",
       "97                 L4       34              {}   \n",
       "99                 L4       50              {}   \n",
       "113                L5       73              {}   \n",
       "118                L4       50              {}   \n",
       "121                L4       50              {}   \n",
       "152                L4       50              {}   \n",
       "153                L4       50              {}   \n",
       "156                L4       50              {}   \n",
       "...               ...      ...             ...   \n",
       "1463              All       73              {}   \n",
       "1464              All       20              {}   \n",
       "1467              All       73              {}   \n",
       "1472              All       73              {}   \n",
       "1493              All       10              {}   \n",
       "1498            upper       64              {}   \n",
       "1512            upper       64              {}   \n",
       "1514            upper       64              {}   \n",
       "1515            upper       29              {}   \n",
       "1533            lower       44              {}   \n",
       "1537            lower       29              {}   \n",
       "1542            lower       29              {}   \n",
       "1567            upper        9              {}   \n",
       "1568            upper       49              {}   \n",
       "1570            upper       49              {}   \n",
       "1579            upper       69              {}   \n",
       "1595            upper       69              {}   \n",
       "1598            upper       49              {}   \n",
       "1603            upper       49              {}   \n",
       "1608            lower       84              {}   \n",
       "1613            lower       84              {}   \n",
       "1614            lower       84              {}   \n",
       "1621            lower       69              {}   \n",
       "1626            lower       91              {}   \n",
       "1627            lower       84              {}   \n",
       "1634            upper       69              {}   \n",
       "1641            lower       84              {}   \n",
       "1643            lower       69              {}   \n",
       "1648              All       69              {}   \n",
       "1651              All       84              {}   \n",
       "\n",
       "                                         markers_absent  \n",
       "8     { Gm13629, Kcnab1,  Fam19a1,  Galntl6,  Endou,...  \n",
       "15                     {Mybpc1,  Npy2r,  Slc44a5, Car4}  \n",
       "26                      { Pter, Ntrk3,  Slc30a3, C1ql3}  \n",
       "27                                           {AI467606}  \n",
       "33                     {Mybpc1,  Npy2r,  Slc44a5, Car4}  \n",
       "35                                               {Car4}  \n",
       "36     {Mybpc1,  Car4,  Npy2r,  Slc44a5,  Sorl1, Npy2r}  \n",
       "37     {Mybpc1,  Npy2r,  Car4,  Slc44a5,  Sorl1, Npy2r}  \n",
       "40                                      {Car4, Runx1t1}  \n",
       "43                                               {Car4}  \n",
       "50    { Gm13629, Kcnab1,  Fam19a1,  Galntl6,  Endou,...  \n",
       "59     {Mybpc1,  Npy2r,  Car4,  Slc44a5,  Sorl1, Npy2r}  \n",
       "60                                           {AI467606}  \n",
       "73                                              {Rspo1}  \n",
       "78                    {Scnn1a,  Fam19a1,  Kitl,  Endou}  \n",
       "80                    {Scnn1a,  Fam19a1,  Kitl,  Endou}  \n",
       "83                    {Scnn1a,  Fam19a1,  Kitl,  Endou}  \n",
       "85                    {Scnn1a,  Fam19a1,  Kitl,  Endou}  \n",
       "88                                                   {}  \n",
       "90                    {Scnn1a,  Fam19a1,  Kitl,  Endou}  \n",
       "92                                                   {}  \n",
       "96                    {Scnn1a,  Fam19a1,  Kitl,  Endou}  \n",
       "97                                                   {}  \n",
       "99                    {Scnn1a,  Fam19a1,  Kitl,  Endou}  \n",
       "113                                      {Fam5c, Ntrk3}  \n",
       "118                   {Scnn1a,  Fam19a1,  Kitl,  Endou}  \n",
       "121                   {Scnn1a,  Fam19a1,  Kitl,  Endou}  \n",
       "152                   {Scnn1a,  Fam19a1,  Kitl,  Endou}  \n",
       "153                   {Scnn1a,  Fam19a1,  Kitl,  Endou}  \n",
       "156                   {Scnn1a,  Fam19a1,  Kitl,  Endou}  \n",
       "...                                                 ...  \n",
       "1463                                     {Fam5c, Ntrk3}  \n",
       "1464  { Rnd3,  Gm13629,  Rcan2, Kcnab1,  Zmat4,  Gal...  \n",
       "1467                                     {Fam5c, Ntrk3}  \n",
       "1472                                     {Fam5c, Ntrk3}  \n",
       "1493  { Gm13629, Kcnab1,  Fam19a1,  Galntl6,  Endou,...  \n",
       "1498   {Mybpc1,  Npy2r,  Car4,  Slc44a5,  Sorl1, Npy2r}  \n",
       "1512   {Mybpc1,  Car4,  Npy2r,  Slc44a5,  Sorl1, Npy2r}  \n",
       "1514   {Mybpc1,  Car4,  Npy2r,  Slc44a5,  Sorl1, Npy2r}  \n",
       "1515                         {Mybpc1,  Npy2r,  Slc44a5}  \n",
       "1533                                       {Crh,  Pnoc}  \n",
       "1537                         {Mybpc1,  Npy2r,  Slc44a5}  \n",
       "1542                         {Mybpc1,  Npy2r,  Slc44a5}  \n",
       "1567                                 {Cxcl14,  Arhgef3}  \n",
       "1568  { Drd1a,  Tacr3, Tcap,  Cnr1, Col25a1,  Il1rap...  \n",
       "1570                      { Drd1a,  Pcsk2, Tcap,  Cnr1}  \n",
       "1579           {Tac1,  Kirrel3,  Etv1,  Snca,  Zcchc12}  \n",
       "1595           {Tac1,  Kirrel3,  Etv1,  Snca,  Zcchc12}  \n",
       "1598                      { Drd1a,  Pcsk2, Tcap,  Cnr1}  \n",
       "1603                      { Drd1a,  Pcsk2, Tcap,  Cnr1}  \n",
       "1608  { Tacr3, Col25a1,  Kcnd2,  Gap43,  Il1rapl2, E...  \n",
       "1613  {Tac1,  Kirrel3,  Tacr3, Col25a1,  Il1rapl2,  ...  \n",
       "1614  { Tacr3, Col25a1,  Kcnd2,  Gap43,  Il1rapl2, E...  \n",
       "1621                       { Il1rapl2,  Tacr3, Col25a1}  \n",
       "1626  {Tac1,  Kirrel3,  Kcnd2,  Gap43,  Etv1,  Snca,...  \n",
       "1627  { Kcnd2,  Gap43,  Tacr3, Col25a1,  Il1rapl2, E...  \n",
       "1634           {Tac1,  Kirrel3,  Etv1,  Snca,  Zcchc12}  \n",
       "1641  { Kcnd2,  Gap43,  Tacr3, Col25a1,  Il1rapl2, E...  \n",
       "1643                       { Il1rapl2,  Tacr3, Col25a1}  \n",
       "1648           {Tac1,  Kirrel3,  Etv1,  Snca,  Zcchc12}  \n",
       "1651  {Tac1,  Kirrel3,  Tacr3, Col25a1,  Etv1,  Snca...  \n",
       "\n",
       "[255 rows x 10 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_types[df_types['coretype'] == 'Transition']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.value_counts(df_types[df_types['coretype'] == 'Transition']['cluster']).plot.bar()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phenotype Bagging"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>short_name</th>\n",
       "      <th>coretype</th>\n",
       "      <th>primary</th>\n",
       "      <th>secondary</th>\n",
       "      <th>cre</th>\n",
       "      <th>major_dissection</th>\n",
       "      <th>layer_dissectoin</th>\n",
       "      <th>cluster</th>\n",
       "      <th>markers_present</th>\n",
       "      <th>markers_absent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A200_V</td>\n",
       "      <td>Core</td>\n",
       "      <td>f01</td>\n",
       "      <td>0</td>\n",
       "      <td>Calb2</td>\n",
       "      <td>V1</td>\n",
       "      <td>All</td>\n",
       "      <td>81</td>\n",
       "      <td>{ Cxcl14,  Cox6a2, Crispld2,  Tpm2,  Itih5}</td>\n",
       "      <td>{ Car4,  Sorl1, Npy2r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A201_V</td>\n",
       "      <td>Core</td>\n",
       "      <td>f02</td>\n",
       "      <td>0</td>\n",
       "      <td>Calb2</td>\n",
       "      <td>V1</td>\n",
       "      <td>All</td>\n",
       "      <td>80</td>\n",
       "      <td>{ Car4, Cxcl14,  Tac2}</td>\n",
       "      <td>{Mybpc1,  Npy2r,  Slc44a5}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A202_V</td>\n",
       "      <td>Core</td>\n",
       "      <td>f24</td>\n",
       "      <td>0</td>\n",
       "      <td>Calb2</td>\n",
       "      <td>V1</td>\n",
       "      <td>All</td>\n",
       "      <td>70</td>\n",
       "      <td>{ Rorb,  Whrn,  Pde1a,  Sparcl1, Rspo1,  Inhba...</td>\n",
       "      <td>{Scnn1a,  Fam19a1,  Kitl,  Endou}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A203_V</td>\n",
       "      <td>Core</td>\n",
       "      <td>f04</td>\n",
       "      <td>0</td>\n",
       "      <td>Calb2</td>\n",
       "      <td>V1</td>\n",
       "      <td>All</td>\n",
       "      <td>63</td>\n",
       "      <td>{ Nrp1,  Slc18a3,  Pcdh15, Aebp1,  Phlda1,  Pv...</td>\n",
       "      <td>{Crh,  Pnoc}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A204_V</td>\n",
       "      <td>Core</td>\n",
       "      <td>f02</td>\n",
       "      <td>0</td>\n",
       "      <td>Calb2</td>\n",
       "      <td>V1</td>\n",
       "      <td>All</td>\n",
       "      <td>80</td>\n",
       "      <td>{ Car4, Cxcl14,  Tac2}</td>\n",
       "      <td>{Mybpc1,  Npy2r,  Slc44a5}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  short_name coretype primary secondary    cre major_dissection  \\\n",
       "0     A200_V     Core     f01         0  Calb2               V1   \n",
       "1     A201_V     Core     f02         0  Calb2               V1   \n",
       "2     A202_V     Core     f24         0  Calb2               V1   \n",
       "3     A203_V     Core     f04         0  Calb2               V1   \n",
       "4     A204_V     Core     f02         0  Calb2               V1   \n",
       "\n",
       "  layer_dissectoin  cluster  \\\n",
       "0              All       81   \n",
       "1              All       80   \n",
       "2              All       70   \n",
       "3              All       63   \n",
       "4              All       80   \n",
       "\n",
       "                                     markers_present  \\\n",
       "0        { Cxcl14,  Cox6a2, Crispld2,  Tpm2,  Itih5}   \n",
       "1                             { Car4, Cxcl14,  Tac2}   \n",
       "2  { Rorb,  Whrn,  Pde1a,  Sparcl1, Rspo1,  Inhba...   \n",
       "3  { Nrp1,  Slc18a3,  Pcdh15, Aebp1,  Phlda1,  Pv...   \n",
       "4                             { Car4, Cxcl14,  Tac2}   \n",
       "\n",
       "                      markers_absent  \n",
       "0             { Car4,  Sorl1, Npy2r}  \n",
       "1         {Mybpc1,  Npy2r,  Slc44a5}  \n",
       "2  {Scnn1a,  Fam19a1,  Kitl,  Endou}  \n",
       "3                       {Crh,  Pnoc}  \n",
       "4         {Mybpc1,  Npy2r,  Slc44a5}  "
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Tables should now be in final version\n",
    "df_types.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "response = requests.get('http://api.brain-map.org/api/v2/data/query.json?criteria='\n",
    "                        'model::TransgenicLine,rma::options[num_rows$eqall]')\n",
    "cre_ref = pd.DataFrame(response.json()['msg'])\n",
    "cre_ref['stock_number'] = pd.to_numeric(cre_ref['stock_number'])\n",
    "\n",
    "#merge both cre dataframes\n",
    "cre_df = pd.merge(df_cre_mtd, cre_ref,  how='left', \n",
    "                  left_on=['Driver Line','Public Repository Stock #'], \n",
    "                  right_on = ['name','stock_number'])\n",
    "\n",
    "#get rid of redunant/useless columns\n",
    "cre_df = cre_df[['Abbreviation','name', 'id', 'stock_number',\n",
    "                 'transgenic_line_source_name',\n",
    "                 'transgenic_line_type_name',\n",
    "                 'url_prefix', 'url_suffix','description']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Abbreviation</th>\n",
       "      <th>name</th>\n",
       "      <th>id</th>\n",
       "      <th>stock_number</th>\n",
       "      <th>transgenic_line_source_name</th>\n",
       "      <th>transgenic_line_type_name</th>\n",
       "      <th>url_prefix</th>\n",
       "      <th>url_suffix</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Calb2</td>\n",
       "      <td>Calb2-IRES-Cre</td>\n",
       "      <td>177837281.0</td>\n",
       "      <td>10774.0</td>\n",
       "      <td>JAX</td>\n",
       "      <td>driver</td>\n",
       "      <td>http://jaxmice.jax.org/strain/</td>\n",
       "      <td>.html</td>\n",
       "      <td>Enriched in restricted populations within the ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chat</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chrna2</td>\n",
       "      <td>Chrna2-Cre_OE25</td>\n",
       "      <td>182693192.0</td>\n",
       "      <td>36502.0</td>\n",
       "      <td>MMRRC</td>\n",
       "      <td>driver</td>\n",
       "      <td>http://www.mmrrc.org/catalog/getSDS.jsp?mmrrc_id=</td>\n",
       "      <td>None</td>\n",
       "      <td>Enriched in cortical layer 5. Also expressed i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ctgf</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cux2</td>\n",
       "      <td>Cux2-CreERT2</td>\n",
       "      <td>177839004.0</td>\n",
       "      <td>32779.0</td>\n",
       "      <td>MMRRC</td>\n",
       "      <td>driver</td>\n",
       "      <td>http://www.mmrrc.org/catalog/getSDS.jsp?mmrrc_id=</td>\n",
       "      <td>None</td>\n",
       "      <td>Enriched in cortical layers 2/3/4, thalamus, m...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Abbreviation             name           id  stock_number  \\\n",
       "0        Calb2   Calb2-IRES-Cre  177837281.0       10774.0   \n",
       "1         Chat              NaN          NaN           NaN   \n",
       "2       Chrna2  Chrna2-Cre_OE25  182693192.0       36502.0   \n",
       "3         Ctgf              NaN          NaN           NaN   \n",
       "4         Cux2     Cux2-CreERT2  177839004.0       32779.0   \n",
       "\n",
       "  transgenic_line_source_name transgenic_line_type_name  \\\n",
       "0                         JAX                    driver   \n",
       "1                         NaN                       NaN   \n",
       "2                       MMRRC                    driver   \n",
       "3                         NaN                       NaN   \n",
       "4                       MMRRC                    driver   \n",
       "\n",
       "                                          url_prefix url_suffix  \\\n",
       "0                     http://jaxmice.jax.org/strain/      .html   \n",
       "1                                                NaN        NaN   \n",
       "2  http://www.mmrrc.org/catalog/getSDS.jsp?mmrrc_id=       None   \n",
       "3                                                NaN        NaN   \n",
       "4  http://www.mmrrc.org/catalog/getSDS.jsp?mmrrc_id=       None   \n",
       "\n",
       "                                         description  \n",
       "0  Enriched in restricted populations within the ...  \n",
       "1                                                NaN  \n",
       "2  Enriched in cortical layer 5. Also expressed i...  \n",
       "3                                                NaN  \n",
       "4  Enriched in cortical layers 2/3/4, thalamus, m...  "
      ]
     },
     "execution_count": 158,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cre_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyontutils.namespaces import makePrefixes, ilxtr, definition\n",
    "from pyontutils.closed_namespaces import rdf, rdfs, owl\n",
    "\n",
    "class Tasic2015(Genes, phns.Species, \n",
    "                phns.Regions, phns.Layers):\n",
    "    branch = devconfig.neurons_branch\n",
    "    \n",
    "    #TODO: Ambiguous, most likely post-clustering layers\n",
    "    L6a = Phenotype(ilxtr.TasicL6a, ilxtr.hasSomaLocatedIn, \n",
    "                    label = \"Tasic Layer VI - A\", override = True)\n",
    "    L6b = Phenotype(ilxtr.TasicL6b, ilxtr.hasSomaLocatedIn, \n",
    "                    label = \"Tasic Layer VI - B\", override = True)\n",
    "    \n",
    "    #Aggregate Layer Phenotypes\n",
    "    with phns.Layers:\n",
    "        upper = LogicalPhenotype(OR, L1, L2, L3)\n",
    "        lower = LogicalPhenotype(OR, L4, L5, L6)\n",
    "        All = LogicalPhenotype(OR, upper, lower)\n",
    "    \n",
    "    #Tasic Hiearchical Cluster Position (1-indexed, breadth first)\n",
    "    cmp  = pred.hasComputedMolecularPhenotype\n",
    "\n",
    "class TasicBagger:\n",
    "    #from Allen Cell Types\n",
    "    prefixes = {**{'JAX': 'http://jaxmice.jax.org/strain/',\n",
    "                   'MMRRC': 'http://www.mmrrc.org/catalog/getSDS.jsp?mmrrc_id=',\n",
    "                   'AllenTL': 'http://api.brain-map.org/api/v2/data/TransgenicLine/'},\n",
    "                **makePrefixes('definition', 'ilxtr', 'owl')}\n",
    "\n",
    "    def __init__(self, data = None, **metadata):\n",
    "        \"\"\"\n",
    "        Initializes a Tasic Neuron Bagger Object\n",
    "        \n",
    "        data: Pandas DataFrame\n",
    "        **metadata: any relevant metadata DataFrames\n",
    "        \"\"\"\n",
    "        self.data = data\n",
    "        self.ns = {k:rdflib.Namespace(v) for k, v in self.prefixes.items()}\n",
    "        if metadata:\n",
    "            for key, dataframe in metadata.items():\n",
    "                setattr(self, key, dataframe)\n",
    "            \n",
    "    @staticmethod\n",
    "    def layer_parse(key):\n",
    "        \"\"\"method that parses the layer labels for Tasic 2015 cells.\n",
    "        returns the layer phenotype.\n",
    "        \n",
    "        key: str, layer label\n",
    "        \"\"\"\n",
    "        if key == 'L2/3':\n",
    "            layer_phn = L23     \n",
    "        else:\n",
    "            layer_phn = Tasic2015[key]\n",
    "        \n",
    "        return layer_phn\n",
    "    \n",
    "    @staticmethod\n",
    "    def gene_parse(gene_set, mode = 'present'):\n",
    "        \"\"\"method that parses a set of gene markers and returns\n",
    "        a list of phenotypes.\n",
    "        \n",
    "        gene_set: set, set of gene markers in dtype str\n",
    "        mode: str, \"present\" or \"absent\" markers.\n",
    "        \"\"\"\n",
    "        gene_phns = []\n",
    "        undef = set()\n",
    "        f = Phenotype if mode == 'present' else NegPhenotype\n",
    "        for gene in gene_set:\n",
    "            if gene in Genes.__dict__:\n",
    "                gene_phns.append(f(Genes[gene]))\n",
    "            else:\n",
    "                undef.add(gene)\n",
    "        if len(undef) > 0: #FIXME: some genes cannot be mapped\n",
    "            mappings, to_add, errors = ncbigenemapping(undef, \n",
    "                                        return_errors = True)\n",
    "            for gene_name, iri in mappings.items():\n",
    "                gene_phns.append(f(iri, ilxtr.hasExpressionPhenotype,\n",
    "                                   label = gene_name, override = True))\n",
    "        if len(gene_phns) != len(gene_set):\n",
    "            print(errors)\n",
    "        return gene_phns\n",
    "    \n",
    "    @staticmethod\n",
    "    def cluster_parse(pos):\n",
    "        \"\"\"Tasic Computed Gene-based hiearchical cluster\n",
    "        pos: int\n",
    "        \"\"\"\n",
    "        cluster_phn = Phenotype(\"ilxtr:cluster\" + str(pos), cmp, \n",
    "                                label = str(pos))\n",
    "        return cluster_phn\n",
    "    \n",
    "    \n",
    "    def transgenic_parse(self, row):\n",
    "        phenotypes = []\n",
    "        pred = 'ilxtr:hasDriverExpressionPhenotype'\n",
    "        cre = self.cre_metadata[self.cre_metadata['Abbreviation'] == row['cre']]\n",
    "        cre_phn = Phenotype('PR:000013502', pred.hasExpressionPhenotype)\n",
    "        ##WHERE R THE IRIs ??\n",
    "        #for tl in cell_line['donor']['transgenic_lines']:\n",
    "            #prefix = self.cre_metadata['Driver Line']\n",
    "            #suffix = self.cre_metadata['Public Repository Stock #'] if tl['stock_number'] else str(tl['id'])\n",
    "            #line_names = []\n",
    "            #if prefix and suffix and prefix in ['AIBS', 'MMRRC', 'JAX']:\n",
    "                #if prefix == 'AIBS':\n",
    "                   # prefix = 'AllenTL'\n",
    "               # iri = self.ns[prefix][suffix]\n",
    "               # phenotypes.append(Phenotype(iri, pred))\n",
    "        return cre_phn\n",
    "    \n",
    "    def build_transgenic_lines(self):\n",
    "        triples = []\n",
    "        for ind, tl in self.cre.iterrows():\n",
    "            _id = tl['stock_number'] if tl['stock_number'] else tl['id']\n",
    "            prefix = tl['transgenic_line_source_name']\n",
    "            line_type = tl['transgenic_line_type_name']\n",
    "            if prefix not in ['JAX', 'MMRRC', 'AIBS']:\n",
    "                print('WARNING:', 'unknown prefix')\n",
    "                continue\n",
    "            elif prefix == 'AIBS':\n",
    "                prefix = 'AllenTL'\n",
    "\n",
    "            _class = self.ns[prefix][str(_id)]\n",
    "            triples.append((_class, rdf.type, owl.Class))\n",
    "            triples.append((_class, rdfs.label, rdflib.Literal(tl['name'])))\n",
    "            triples.append((_class, definition, rdflib.Literal(tl['description'])))\n",
    "            triples.append((_class, rdfs.subClassOf, ilxtr.transgenicLine))\n",
    "            triples.append((_class, ilxtr.hasTransgenicType, ilxtr[line_type + 'Line']))\n",
    "\n",
    "        # TODO aspects.ttl?\n",
    "        transgenic_lines = simpleOnt(filename='tasic-transgenic-lines',\n",
    "                                     path='ttl/generated/',\n",
    "                                     prefixes=self.prefixes,\n",
    "                                     triples=triples,\n",
    "                                     comment='Tasic transgenic lines for cell types',\n",
    "                                     branch=self.branch)\n",
    "\n",
    "        transgenic_lines._graph.write()\n",
    "        \n",
    "        #return transgenic_lines   \n",
    "    \n",
    "    @property\n",
    "    def bags(self):\n",
    "        for row in self.data.itertuples():\n",
    "            with Tasic2015:\n",
    "                #Every neuron sampled in this paper is from V1\n",
    "                with Neuron(Mouse, V1) as context:\n",
    "                    label = str(row.short_name) #change this\n",
    "                    cluster_phn = Phenotype(str(row.cluster), cmp)\n",
    "                    layer = layer_parse(row.layer_dissectoin)\n",
    "\n",
    "                    present_phns = gene_parse(row.markers_prsent, mode = 'present')\n",
    "                    absent_phns  = gene_parse(row.markers_absent, mode = 'absent')\n",
    "                    markers_phns = present_phns + absent_phns\n",
    "\n",
    "                    phenotypes = [layer_phn, cre_phn, cluster_phn] + markers_phns\n",
    "            \n",
    "            yield label, phenotypes\n",
    "          \n",
    "    \n",
    "class TasicNeuron(NeuronEBM):\n",
    "    owlClass = ilxtr.NeuronTasic2015\n",
    "    shortname = 'Tasic2015'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = {\"cre\":cre_df}\n",
    "\n",
    "def main():\n",
    "    ttl_test_path = '/var/host/media/removable/SD Card/Neuron/Tasic/ttl_export'\n",
    "    config = Config(\"tasic-2015\", ttl_export_dir=Path(ttl_test_path))\n",
    "    tb = TasicBagger(data = df_types, **metadata)\n",
    "    for label, bag in list(tb.bags):\n",
    "        TasicNeuron(*bag, label = label, override=True)\n",
    "    config.write()\n",
    "    config.write_python()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: unknown prefix\n",
      "WARNING: unknown prefix\n",
      "WARNING: unknown prefix\n",
      "WARNING: unknown prefix\n",
      "WARNING: unknown prefix\n",
      "WARNING: unknown prefix\n",
      "WARNING: unknown prefix\n",
      "WARNING: unknown prefix\n",
      "WARNING: unknown prefix\n",
      "WARNING: unknown prefix\n",
      "WARNING: unknown prefix\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'TasicBagger' object has no attribute 'branch'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-170-f305fa9ef066>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mmetadata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m\"cre\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0mcre_df\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mtb\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTasicBagger\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf_types\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mmetadata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mtb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbuild_transgenic_lines\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-169-9d9940826102>\u001b[0m in \u001b[0;36mbuild_transgenic_lines\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    132\u001b[0m                                      \u001b[0mtriples\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtriples\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    133\u001b[0m                                      \u001b[0mcomment\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'Tasic transgenic lines for cell types'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 134\u001b[0;31m                                      branch=self.branch)\n\u001b[0m\u001b[1;32m    135\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    136\u001b[0m         \u001b[0mtransgenic_lines\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_graph\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'TasicBagger' object has no attribute 'branch'"
     ]
    }
   ],
   "source": [
    "metadata = {\"cre\":cre_df}\n",
    "tb = TasicBagger(data = df_types, **metadata)\n",
    "tb.build_transgenic_lines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test Code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Neuron(Phenotype('NCBITaxon:10090',\n",
      "                 'ilxtr:hasInstanceInSpecies',\n",
      "                 label='Mus musculus'),\n",
      "       Phenotype('UBERON:0002436',\n",
      "                 'ilxtr:hasSomaLocatedIn',\n",
      "                 label='primary visual cortex'),\n",
      "       Phenotype('UBERON:0005395',\n",
      "                 'ilxtr:hasLayerLocationPhenotype',\n",
      "                 label='cortical layer VI'))\n"
     ]
    }
   ],
   "source": [
    "with Tasic2015:\n",
    "    with Neuron(Mouse, V1) as context:\n",
    "        a = Neuron(phns.Layers.L6)\n",
    "        print(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyontutils.core import simpleOnt\n",
    "simpleOnt()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Phenotype(\"ilxtr:cluster6\", \"ilxtr:hasComputedMolecularPhenotype\", label = \"Tasic\", check = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyontutils.utils import byCol, relative_path\n",
    "from pyontutils.namespaces import ilxtr\n",
    "from pyontutils.closed_namespaces import rdfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with Neuron(phns.Species.Mouse, phns.Regions.V1) as context:\n",
    "    print(context)\n",
    "    n11 = Neuron(phns.Layers.L1)\n",
    "    print(n11)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
